{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c114f17c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyjaspar import jaspardb\n",
    "from Bio import motifs\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import csv\n",
    "\n",
    "# Fetch motifs from JASPAR\n",
    "def fetch_motifs_from_jaspar(tf_list, release=\"JASPAR2024\"):\n",
    "    jdb = jaspardb(release=release) # set the JASPAR object\n",
    "    motif_dict = {} # Create empty dict\n",
    "\n",
    "    # Iterate through transcription factors\n",
    "    for tf in tf_list:\n",
    "        motif_list = jdb.fetch_motifs_by_name(tf) \n",
    "        if motif_list:\n",
    "            jaspar_motif = motif_list[0] # Store 0 postion of motif_list to variable\n",
    "            motif_dict[tf] = jaspar_motif # Add to dict\n",
    "            print(f\"Fetched motif for {tf}: {jaspar_motif.matrix_id}\") # Print TF and TF id to confirm\n",
    "        else:\n",
    "            print(f\"No motif found for {tf}\") # Print this message when there is no motif information.\n",
    "\n",
    "    return motif_dict\n",
    "\n",
    "# Scan promoter regions using both orientations of motif on forward strand\n",
    "def scan_promoters_for_motifs(fasta_file, motif_dict, threshold=7.0):\n",
    "    results = []\n",
    "\n",
    "    # Parse the fasta file which contain promoter sequences\n",
    "    for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "        gene_info = record.id # Extract the gene information\n",
    "        promoter_seq = str(record.seq) # Extract the sequences\n",
    "\n",
    "        # Extract the gene_name and gene_id from fasta file\n",
    "        if \"_\" in gene_info:\n",
    "            gene_name, gene_id = gene_info.split(\"_\", 1)\n",
    "        else:\n",
    "            gene_name, gene_id = gene_info, \"NA\"\n",
    "\n",
    "        # Iterate through motif_dict\n",
    "        for tf, motif in motif_dict.items():\n",
    "            motif_len = len(motif)\n",
    "\n",
    "            # Get both forward and reverse PSSMs\n",
    "            pssm_fwd = motif.pssm\n",
    "\n",
    "            # Search with forward PSSM\n",
    "            forward_matches = list(pssm_fwd.search(promoter_seq, threshold=threshold)) # Scan Promoter sequences with forward motif patterns\n",
    "            for pos, score in forward_matches: # Iterate through matches\n",
    "                start = pos \n",
    "                end = pos + motif_len\n",
    "                if start < 0 or end > len(promoter_seq):\n",
    "                    continue # Skip when start value is negative or end value is over the promoter sequence length.\n",
    "                results.append([gene_name, gene_id, tf, \"+\", start, end]) # Add matched data into empty list to write CSV file\n",
    "           \n",
    "    return results\n",
    "\n",
    "# Save results to CSV\n",
    "def write_results_to_csv(results, filename=\"motif_scan_results_forward_only.csv\"): \n",
    "    with open(filename, \"w\", newline=\"\") as csvfile:\n",
    "        writer = csv.writer(csvfile)\n",
    "        writer.writerow([\n",
    "            \"Gene_Name\", \"Gene_ID\", \"TF\", \"Strand\", \"Start\", \"End\"\n",
    "        ]) # Write header row\n",
    "        writer.writerows(results) # Write the rows in CSV file\n",
    "    print(f\"Results written to {filename}\") # Print this message to confirm\n",
    "\n",
    "# Run functions\n",
    "if __name__ == \"__main__\":\n",
    "    tf_names = [\"IRF3\", \"IRF7\", \"IRF9\"] # Set the list to fetch multile transcription factors\n",
    "    fasta_file = \"Promoter_sequences.fasta\" # Sequence data to scan\n",
    "\n",
    "    motifs_dict = fetch_motifs_from_jaspar(tf_names) # Store function result in variable\n",
    "    scan_results = scan_promoters_for_motifs(fasta_file, motifs_dict, threshold=3.0) # Store function result in variable\n",
    "    write_results_to_csv(scan_results)"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
