{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0923b801",
   "metadata": {},
   "source": [
    "## Retrieve the promoter sequences ##\n",
    "#### IFNB1, IFIT1, MX1, TLR7 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "ed827432",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/sang/Library/Python/3.9/lib/python/site-packages/urllib3/__init__.py:35: NotOpenSSLWarning: urllib3 v2 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with 'LibreSSL 2.8.3'. See: https://github.com/urllib3/urllib3/issues/3020\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">IFNB1_ENSG00000171855\n",
      "GTAGTTTCTTTTTTCTTTGAGAGCCATCATCACCATCATGGTTGACACCATGAACCTATCTGAAGATGTCAGCCATAGACTGCTTGATATTCTACAGGAAAGATCACAGTTTTAAGTGCAATCTACCCATGTTATTAGCAGTGTGTATCTTTCACACATTACACAGCCTCTCTAAGCCTCATTTCTCTCCTCTGTAAGATGGGGATGATAATAACCCATCTCAAATGTTTACTATGAGGATTATTCAAAGAATGGCAAATAGCAAGTGCTTAATAAATGATAACTAGTACTACCGCCACTACTGTTGTTTTTATTGTATTAGATTATGAACTCTCTAAGGACCATTTCCGGATGGAGGATAAGAGACCATTTGATGTGGGCAGTGATGAGGCCTTCTGTTGCACCTGGAAAGGTCAACTATATACAAGCCTGCAAGTCATTCTATAGGAGCAGGCCCCAGTGACCAGACTCTATAGACTGTCTCCTCTTTCCTGAGAGGGACAGCCATCTCTAGGTTGACTAACCTCTGAAGCTCCTTGCATTGGCTTTTGTGCTATGAGCCATGGATGATTCCAGACTAATCCGAGAATGCTCGTCAAAACCCCAAGGAATTACTCAAATACTGACATAACAGACATTTTTGAGTGGAAGAGCCGAGTTTTTTTTAATATTCTGAAACTCATTGTTTTTAAAATGCATGAGATGGCCAAGGTCTTGCTAAGAGCTGGCCTGCAAAGGCCAAAAGGCCAGAGAGAATGAAACCCATAGAGAGGCAGAATAACCAGAAAGGTTGGGACTCGTTTATTTTATAATGTAAATTAGTCTATTATGAAACAATACTTGTTTACTGGTGGAAAATTGGAAAATACAAAGAATAAAAGGAGGAAAAAAATCACTCTTTAGTTTCACAAGCCAAATCAAGCCACTATTAAAATGGTGGTTTACTTCCTTTTATTAATTTTCTCTACATATTTTTGCATAATCATGTTGTATGTACAATTTTATGTTCTATTTTTCAATATTAACTGGTGTCTTTCAAATTTCCTAATGACAAAAATAATATATGCTCATAATAGAACATTTTAAATGCAAATAAAACAAAATAAATGTTAAAATTTAGTAATATTTATTAAATTTTCTCCAAGTGCACGAAATTACAAATGTAACAACCTAATTCCCTAGTGGCCTAATAACCCTATTTCCCAGACCTCTTCTCATTACAAGGAAAAACTCATATGCAGATAGTTCTAAAGGTATGAAGTGAAAAGATAAAGATTTTTCTTCCTTGCTGCATCCTCACCCCATCAGCATTATTCCCCAGGGTAACTACTATTAATAGATAGTAATTCTACCCAAAGGAAAAAATCATATGCATATAACAGCATCATATGTATACCTTTCTAGTAACTTACAAAACAAATGATAATATCATATCCTTTCTTATGTGTATTGCTCTTTTCACTAAATGTATCTGTGATATGTGTCTATATCAGCTGATTGTCCTTTTTGATGGCTGAATAATATTCCATCTTGTCCACGTGATAGTATTACTTGACAAGCTCCCTGCTGATGGACATTTGTCTTTGTTACTATGATAGTAATATAATCAACATTTATATATGTTTTGTATGTATCTATAATACACATGCACATACACATGCATATTTCTGCAGGGATAGCCATAGTAAATAACTAGTAACGGTATTGCAAGTTAAAGGAACAATCTCATTGCTTGAAATTTTAAATTTTGAAATACACTGCCAATTTTCATGGTCTCTCCTTGTAAGCTAGTTTGGGCTTTCTCACAGCATGACAGGCTCAGGGCAGTCAGACCATCCTGGCCAAAGAGCAGAGTGCCACAGACCACAACTGCTTCTAATCAGCCATCTTCCCAAAGCCTTCTCTTTTTTCTATTAATAACTTTGTATGAGATTCCATCTTAATACTTTTCTGTTGTTTGGTCTTGTAAGAGCTTATTTTTCCTGAACCAGGAAGTGGTTCAGGGCGGTTTTTCTAACTTCACAGAGCTCCCTCTTCTGTTAGCTTTTGTGAAATGGTCAAAAACATAGCAGCCTGCCTTCTGAGTTCTCCATCCCACCCTGGTTGGGCCTTCTCTATCCTTGTCTGTGTTGTTTATATCCTGCTGAAGTGTGATTCCACTTGTGGCAGTTTCTCCTCTGTGTAGGATCAAAAGGGCTGTGACTGGTTGGTTTGAAAATTTCTTATACCCTAGACTATTCCAGTGCCTTTCAGAAGTTTCCAAGGCCCTCTCACACTAATCTATTATCATATTGGGCAAAACTCCTTGCAGTTTCAGCTACTATTCCCTGATTGACTTTTCAGTAAATCTATCTCTCAGTCTTTCAGTATCCAAAGAAGATTGGTTCTAGGACCACCATCCCGCTGCCTCCACAGATACCAAAATCAGAGGATGCTCAATTCCCTCTTATAAAACGTTGCAGTATTTGCATATAATCTGCACATGTATTTCTGTATATTTTAAATCATCCCTAGATTACTTATAATACCTGATACAATATAAATGCTAAATAGCTGTAACACTGTATCTTTAAAATTTACATTATTTTTTGTTGTTGTATTATTATTTTTATTGTATTTTTAAAAAATATTTTCCATCTACAGTCAGTAGAATCCACGGATACAGAACCTATGGATAGGAAGGACCAACTGTATCTTTTAGTGTTTTGAGGTTCTTGAATTCTCAGGTCGTTTGCTTTCCTTTGCTTTCTCCCAAGTCTTGTTTTACAATTTGCTTTAGTCATTCACTGAAACTTTAAAAAACATTAGAAAACCTCACAGTTTGTAAATCTTTTTCCCTATTATATATATCATAAGATAGGAGCTTAAATAAAGAGTTTTAGAAACTACTAAAATGTAAATGACATAGGAAAACTGAAAGGGAGAAGTGAAAGTGGGAAATTCCTCTGAATAGAGAGAGGACCATCTCATATAAATAGGCCATACCCATGGAGAAAGGAC\n",
      "\n",
      ">IFIT1_ENSG00000185745\n",
      "GCTGATAAGAGATTTTAGACTACAAGAGATGACCCTGGAAAAGACAGGGGAAATTTCTATGGAGCCCATTCCTTAATAAGTTGTGATTGGATATTTGAGACCCTTCAGCAGAAATGGCTTAGCATTTGTCCATATTATCATATTATAATTTTTAAAAAATACACTCAGAATAACATTCTTTTTCCATATCATCCTGTGAATGAAGAAGAGACTAAGGAATTCATTACCCATTGTGTAGGTGAGGAAACAAGGCATTGAGTCGTAAAATACTGAATCTGGGCCTCACAGAACAGGAAGTGGGTCCAGGGATTGAGTCCCATATCTCTCAGTCAGAGAATTTCAGACATACAATAGACATTAGGGATAATTGAGGCCCTGGTTCCCGGCTCAGGGCGCAGGTCTACATCAAAATCACCTGAAGATTTCTTTCTTTTTTTTTTTTTTTTTTTTTTTGAGATGGAGTTTCACTCTTGTTGCCCAGGCTGGAGTGCAATGAAATGGCACGATCTCGGCTCACCACAACCTCCACCTCCCGGGTTCAAGCGATTCTCCTGCCTCAGCCTCCCGAGTAGCTGGGATTACAGGCATGTGCCACCACGCCCGGCTAATTATGTATTTTTAGTAGAGACGGGGTTTCTCCATGTTAGTCAGGCTGGTCTCGAACTCCCAACCTCAGGTGATCCGCCCGTCTTGTCCTCCCAAAATGCTGGGATTACAGGCGTGAGCCACTGTGCCCAGGCTGAAGATTTCTTTTTAATACTGAGCCAGAGGCCTGTACCTAAGTATATGTTCAATTGTTCTGGAATGGGCATAGTTAGCATTATTTTTGAGATTTACCTGGCTGATTCTAATGTGCAGTGAGGGTTGAGATCTGCCAACCTATTCCAAATGACTTCTCAGCTGAGGAAATTGAGATTCATAGAGGTGAAGCAGTTTATCATAACTGGTCAGTCCCTTGCAGAGCCCAGCTTAGACTCCTGGGAATGAACTCACAACCTAGTGCTGAAATTTCATTTTCTAAATGTTTGATCCTCTCTGGTTCCCATTTTCTTTTTTATCTGTAAACACACACACACACGCGCGCGATTGGCTTTTCTTTCCTTAATGCTGTTCTATCCTGTGTCCCTCATGGGTTTGTTTAGCCCAAAGAGACACTAAGCAAGATTCTGAATACCTAATCAGCCTGGGCCATAGCCCCGTAGAACCAAGACTGCTGCCAAATTCACAGCAATGGCAAGATAAACTTTTTGTTCATATCTCTGCATGCTATCTAGTGCATTGGCAAATTGAAAACCCATTTTTTCCCATAAGAATTCCAAAGACATCTTGCTCCCCAAAAGATGTTTTCTAATTCATCTGTTTCAGCATTTGCTTATGGCCTGATCATGTATCTGAGTGCAAGACAAAACCTGTCCTATTAATTCAGAAATTCTTATTTGTGTAGAACTTTCTAACTAGTTGTTGAACTCTCTGTGGAAGCTTACAGAATTGGTGAGAGCTGACATGGGAGATGAGTCTGGCAGGGAGTTGAGCAGAAACTGTTTAGATCTGGCAGTGTGCCCAGTTAAGTGTTTCCAAAAGGAAAATACTCTTGCACCTGCTCCTCTTCTGCTTCTTATGTTAAAAGACAAAGTAGGTAGGTTTTTTTAAAGGAAACATTGTTTTTATATATTGTTATGATTATTTGAGACAGGGGCTCACTCTGTTGCCCAGGCTGGAGTGCAGTGGTGTGACCATGGCTCACTGCAGCCTCTACCTCCTGGGCTCAAGTGAGCCTCCCACCTCAGCCTCCTGAGTAGCTGAGACCACAGACACACACCACCACACCCAGCTTTTTTTTTCAGTTTTTTTGTAGAGTTGGGGTCTCCCTATGTTGTCCAGGCTGGTCTACAATTCCTGGGCTCAAGTGATCCTCCCGCCTTGCCTTCCAAAATACTGGGACTTACAGGCCTGAGCCAATGCACCCGGACAACAATTTTTTATTTTTTATTTATTTATTTTTTTAGACGGAGTCTCGCTCTGTCACCAGACTGGAGTACAGTGGTGTGATCTCGGCTCACTGCAACCTCTGCCTCCCAGGTTCAAGCAATTCCCCTGCCTCAGCCTCTCGAGTAGGTGGGACTACAGGTGCACACCACCACACCCAGCTAATTTTTTGTATTTTAGTAGAGAGGGGGTTTCACCATGTTGGCCACGATGGTCTCCATCTCCTGACCTTGTCATCCGCCCACCTTGGCCTCCCAAAGTGCTGGGACTACAGGCATGAGCCACCGCACCCAGCCAAGAATCATTATTTTTAACTTGATGACTGAAAATAATAATAATAATAGTTACCACTTATTTGCATGCTTCTATGTGCCAGGTAGTTGCTAACTATTTAAACTCAAATTCCATGAACTGTAGTGGAGGTTGTACTGGAATTTGATTCAGAATGACAGTGTCCATGATGGAGCAATAGAGGGCTCTCTATTTCAAACCATACCTCCTTGCTTTTACCTCCTGCCTAAGTCATCAGGGGTTAGAAGGCTTTCTAGGTATTGGTCTCTTTCCTTCATTCCTAAACCAGATTGGTTGCTTATTTCCGTCAAGCTGAAACCAAAAGTAAGCAACCAAAAAGCAACCAGCAACCAAAAGCCTTGTTACTCAATTAATTAAGAGTAGATTTTTATATTTGATAGTAGGTTCCTTCTAAATATAGAAACTGAAAATAGAGCTATCTCCTTCAATTCTCCTTTTTCTGTGTATTCATCCAGAATCCAGCCACCAACTGCCACAATAGGCAGCAATGGACTGATGTTCTTTAGGGAGGACGTGAATCTCGTTCCAAATGCTGGCCAGTCATTGGGTTTCTGCAGCACTAGAAACATCTATGGTTGCAGGTCTGCAGTTTATCTGTTTTAAAATAGAAACAAAGTTTCATTCCCCACCCCCCCCCGTCAGCAGGAATTCCGCTAGCTTTAGTTTCACTTTCCCCTTTCGGTTTCCCTAGGTTTCCAACTT\n",
      "\n",
      ">MX1_ENSG00000157601\n",
      "AATGAAATAATGGCATTCACAGCAACCTGGATGGAGTTGGAGACCATTATTCTAAGTGAAGTAACTCGGGAATAAAAACCAAACATTGTATATTCTCACTCATAAGTGAGATTATACAAGCTATGAGGATGCAAAGTCATAAGAATGATACAATAAACATTGGGAACTCAGGGGGAAGGGTGAGGGTGGGGTGAGGGATAAAAGACTAGACATCGGGTACAGTATACACTGCTTGGGTGATGGGTGCACCAAAATCCCAGAAATCACGATTAAAGAACTTGTCAATGTAACCAAACACCACCTGTTCCTCAAAAACTATTGAAATAAAAAATAAATACATAAAAGAAGTTGAGGTTAAATTAATCTAAGTAGGGAGTGCATTTGGGCCAAGCTTGAGGATTGCAACGTTGGAGCACAGATTTAAGTTGCCCTGAATCTACACTCTGATAAGTGGCATTTATAAGTGGATTTTTAAAGGCAAAAGGGGGAGACAGGGAGAGGGCTGATACAAAGCTGTTTGTCAGGAATTCTTATTGGTTTACAGAAATAACATTGATAAGTGCTTATTAGTCCATTTTCATGCTGCTGATAAAGACATACCTCAGACTGGGCAATTTACAAAAGAAAGAGGTTTATTGGACCTATAGTTCCACATGGCTGAGGAGGCCTCACAATCATGGTGGAAGGCAAGGAGGAGCAAGTCACATCTTACATGGATGACAGCAGGAAAAAAGAGAGAGCTTGTGCAGAGAAACTCCCTTTTTTAAAACCATCAGATCTTGTGAGACCCGTTCAGGATCATGAGAACAGCACAGGAAAGACCCACCCCCAAGATTCAATCATCTCCCACCGGGTCTCTCCCACAACATGAGTGAATTATGGGAGCTACAGGATGAGATTTGGGTGAGGACACAGAGCCAAACCATATCAGAGTGATTGACTATACATTGTTAAGCTACAGGGTGTGGATTACAGCGTCCAGTGTGGCATTATTAGGTTAATCCATAGCTATTAATACCTGTGGCAATAGTAAGAGGTTTCAAGAGATGAACACACAGCTGAAAGTGGGGAGTAGGGAGTGGTTGCTGTCTTATTTTAATGTCTCTCTGGGCCTGATAATTTTAAAAGCCTTGCATTCTTCGGATACAAGTTCCTTTCTTTTTTTCAAGAACAACAGAGAAAGGCTGCCTGGCTACCTTTACCTGCAGCACCTCTGATAGTGGCTGCTTGGCAGCTGGTTAGGTAAACCCAATTTCTACTGGATAAACTTCATGGACATTCACCCATGTTAGTACTGGGGAATGCTCTGAAATAGCAAACTTGTCCATACAGCTCCCTTGCTGCTCAGAGGGTCTGGAGAGGAACAGCAGAGGCCCAGCCCCAAATACTTCTTGTACCGGCTTCTATTATGAGAGGTGATGGCACCTTTAGGCACTTATATTTCAGGTAGAAGGAGCAGGAATAAAGAAGAGAGGGCCACAATCTGAAATATCAGTTCTGGATCTCCCTGGGGGCAAGACACTGGCCCTTAGCTGAGCTATTCTGTCATCACTGTGAGCCCTCATGCTCGACCTCACCGGAGCACCCCTCGATCATGTGCTGAATGCCTAGCACGCGCCTGGCACTGTGGGGTACAGAAATCTAGAGCACAGAGCTGAGTTTTCAGGGCTTAGCTGTCCTCACATCTGTGACACAAACAAGGATCAAGGAAAGTTTTGTTTTTGTTATTTTTATTTTTTTTTAAGTTGGGGCCTTACTCTGTAGCCCAAGCTAGAGTGCAGTGGTGCAAACATAGCTCATTGCACCCTCGAACTCCTGGGCTCAAGTCATCCTCCCACCTTGGCCTCCAGAGTAGCTGGAAGCCCAGCTCAAGGAAAGCATTAACATGTTCTACTTTTCACCATAACTCCTGTGTATTGGGAAATGGAAAATGAGCTTTTGAAGTGGCCTATTCGCTTCCTCCATGAGTACTTCCTGAACACTTTCCTCTTGGCCACTGCTCCCAAGACCCCCTCAAAGTCTCTCTCCTGGACTTCCAGCTTCTGGAATGAGCCCCAAGAGTGGCACTTAGCTCCAGCTCCTAGTAGCTCACTCCTTCCTTGCTCACCTGTGCCCCCCACCTGTCTCCTTAGGGGCAGTGGCTTTGTCTTACTCATTTGCATATCACAAATGTCCAGCATAGAAGCCAGCATGGGGCCCAGTAAGTGTTTGTGGAGAAAGAAAGGGAAGAGGTGAAGAAGGGAGGAGACGGAGGGAGAAAAGGAGGGCAAGAGAGGGGATGTCATGTCATGTCATGTCATGTCATGTCAGTTCTCCCTTCCTAGATCTGGACAAGGGAAAGAAAAGCAAGCAAGCCTGACCGCCTCAGTGCAGACCCTCTAAGATGCAAGCCTTCCCCAGGTAAGACAGTTCCGCAGGGTGCTTCTGTGGGCTGGCTTTCTGAGCAGCCATCTCAAAGTATGCGAAAGAAATAGATTTTGGGCCGGGCACGGTGGCTCATGCCTGTAATCCCAGCACTTTAGGAGGCCGAGGCGTGTGGATCACCTGAAGTCAGGAGTTCAAGACCAGCCTGGCCAACATGGTGAGACCTTGTATCTACTAAAAATACAAAAACTAGCCGGGCTTGGTGGCGCATGCCTGTAGTCCCAGCTACTCAGGAGGCTGAGGCAGGAGAATCGCTTGAACCCAGGAGGTAAATGTTGCAGTAAGCCGAGATTGCCCCACTGCACTCCAGCCTGGATGACAAGAGTGAGACTCCATCTCAAAAAAAAAAAAAAAAAAAGATTTTAGGGGTAAAATATTTTGCTTTCCTTCAGTGGCTGGGCCACCTGTACACAATGCCTGAACCCAGGTCCTGCAGCCCCCAGCTTGGCTCTCCAAAGCTCACCAGTATCAAAGGATTTCAGCTTCTGCTCCTCTCCGCAGGGCTCTCTGGTGGGCCCCAGACAGAGATGAAGGAAAAGTCTGGCCTTGACCTTGAGGACCAAAAGCGACCACGTC\n",
      "\n",
      ">TLR7_ENSG00000196664\n",
      "ATGGGCGTGGGACCCTCCGAGCCAGGCGCGGGATATAATCTCCTGGTGTGCCATTTGCTAAGACCATTGGAAAAGCACAGTATTAGGGTGGGAGTGACCCGATTTTCCAGGTGCTGTGTGTCACAGCTTTGCTTGGCTATGAAAGGGAATTCCCTGACCCCTTGTGCTTCCCAGGTGAGGCAATGCCTCGCCCTGCTTTGGCTCATGCTCGGTGTGCTGCACCCACTGTCCTGCGCCCACTGTCTGACAAGCCCCAGTGAGATGAACCCAGTACCTCAGTTGGAAATGCAGAAATCACCCATCTTCTGCATTGCTCACGCTGGGAGCTGTAGACTGGAGCTGTTCCTATTCAGCCATCTTGGAACCACCCTACAGTTTCTTTCAATACTTTTCATGTTCAATGTCTTACACTTAAATTTTTCAACTACATGAAATTTATTTTGATATAAGAAATGAGGTTGGGCCAGGCGCAGTGGCTCACACCTGTTATCCCAAAACTTTGGGAGGCTGAAGTGGGCGAATCATGAGATCAGGAGTTCAAAACCAGCCTGGCCAATATAGTGAAACCCCACCTCTACAGAAAATACAAAAAATTAGCCGAGCGTGGTGGCGGGTGCCTGTAATCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCTGGGAGGTGGAGGTTGCAGTGAGCAGAGATTGCGCCACTGCACTCCAGCCTGGGTGACAGAGTGAGACTACATCTCAAAAAATAAATAAAAAGACAAATGAGGTTGGACAATAGCTTGATTTTTTAAAAACTCCAAGCCAGTTGCCCAATAATCCATTTTGCCATTGATGTGACTTGAAACTAAACACCCATATGTGCTTTTACTTAATTCTGCTCACTTCTCAGTTTCACTGATGCACTGGTACTATGCCATTGGAGGCAATACAGCATTCTAATGCATTTACAGATATGGAAGAGAAATCCTCCCCTTTTATTTTCATTCAGAAATTGCCTGGCTATTCTCACATGTGGTCTTACCCCAGTCAGACCCAACACCTCACTTTTATAACAAATATTTGGTAATGCACCCTTTGTTATATGAAAGGAGATATTTGTGGATAATGTAACCCCAGTCTTCATGATAAACAAAAAGGCCCAGCTGATTTCCAAAATGCACCCCAGTTTAGAATCAGTCTGGCAAGTATCACATGAAATCCTATTGGTATTTGATTGGGATCACACAGCATTTGCAAATCAATTTAAGCAACATTTCTATTTTTACAATATTGTGGCCTCTAGCCCCAATAACAATTATTTCTCTTCATTTACTTATATCTTTGTGTCTGGGCAGGGTCCTGCCAGGAAACAGACGGCATGTTAAAGTGAGAAACTGATGAGTTCAGCAAAGTGACTATTTATATATTTGGACAGGATTTAAGGAAGTAAGAAAGGATGATGCAACACTTCAGAGGGGAGTCTTTCCGACCTCAGGCTGAAGGAGAAGGAACGATTACTGGAATTCAGGGAGGAGAGCATCACCAAACAAGAGCTTCATTAGAGGACTGCAGCCAACGCAGGGCCAGGCGGAGGGAGCCAGAGGGAGGCAGGCTCTGCTCTCCCTCTTCCTGCCCTTCAGTTTCCAACCACGGCCTCCTATTGGCCAAACCCAACCAGAAGCCAGCGAGCAAGGGGCTACTGATGAAGCACATATAGCTCAGCCTCCAGAGACACAGAACAGGATAAAAGGCTGAGACAGTGGGTCTGGTGGGGCAAAGAGAAAGCTTGCACCTGCCAGGTAAAGCATTATAGTCCCCATCCTCCCCCCACCACCTTAGTTCTTGTGCATTTCCCATCAGTTTTCTTCCAAAGCATTTCAGATCTCACTGATTTGAATGGGACCTCTTCTTCTATTACCAATTTGGATCAGTAATTATTTATGTATGAGAAAACTATTGATTTTTACATATCGTTGCATAGCAGAGTGGTTTAATAAGGAGACATTTGGTTTTTACTGCCTGTGTGGGTATCACTTGCTATGTGACTTGAGGCAAATCCAATATTTCTTCTGTTATAAATTCCAGTATTTGTAAAAAATGGGTAATAAGATCTCTATTTTATATAGTTTTAGTATTTAATGAGATAATACATATAAAGTCATTAAAACAGTGTCTGGCTCATAAAAAACCCTCAATAAATGTCACTTATTACTGTATCTGGTTTTTGAGCTGCTCTATTGCACCATTGAGTTTTCAGCCCAGTATATGTTAACCCTGATCATTATCTGCAGAAGTCCCCGTGCCACACTCTACATCATCCAAATTCTCTCCAGGTGGACTAAGTAGATTAAAGAACTTTAAACATAACTACCATATTTTGGCTCTATCTACAAAATGTCCAATAATCAGTTAAGAAAGGAACAATTCTCTTGGGGCCCACACTTTGAGAAGCAAATGCAGCTGAACTTTTTTAGAGGAAAGTGAGTGAACCAACTGGTAGCTTTGCCACTGCTTAAAAACCAGCATCCTTTCCAGCTGGGTCTAAGACAGAATAAGGTAAATTTAGATATGTCTCTAATATATCTATAGAACAGTGGTTCTCAACCCGGGGTGTTTTTGCCCCTTAGGGGATAATTTGCAATGTCTGGAGACATCTGTGATTGTCATAACTGGAAGGGGGCAGTGCTATTGGCATCTAGTGGGTATAGAGCAAGGGTGCTACCAAATATCCTATGGTGCAACAGAGAATTATCTGGTCAAAAATGTAAATAGTGCTGAGGGTGAGAAACCCTGCTATAAAAACGAAAGAAATTTGGTCTACAGAGTTGTTTGGATTTAGACAAGACGTTGCCCCAATAGTGGTGATAGAAATAAGAGGAACCCCGTGCTTTTGCAAAGCCCATATCTGGGGTGGCTTAAATAATCATGCTCCTCCCCATCCCCCGACCTGATCTTTGTAGTTGGAAACTCCAGGGCTGGCTGCCTGTAGTCTTTGTGACTACACTTCCTGCCTCCCATC\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "import sys\n",
    "\n",
    "# Set the server database to retrieve\n",
    "server = \"https://rest.ensembl.org\"\n",
    "headers = {\"Accept\": \"application/json\"}\n",
    "\n",
    "# Set the Gene_names and ids\n",
    "gene_ids = {\n",
    "    \"IFNB1\": \"ENSG00000171855\",\n",
    "    \"IFIT1\": \"ENSG00000185745\",\n",
    "    \"MX1\": \"ENSG00000157601\",\n",
    "    \"TLR7\": \"ENSG00000196664\"\n",
    "}\n",
    "# write a fasta file for promoter sequences\n",
    "with open(\"Promoter_sequences.fasta\", \"w\") as fasta_file:\n",
    "    for gene_name, gene_id in gene_ids.items():\n",
    "        # Set the endpoint to get 5' end expanded sequences\n",
    "        endpoint = f\"/sequence/id/{gene_id}?expand_5prime=3000\"\n",
    "        url = server + endpoint\n",
    "\n",
    "        r = requests.get(url, headers=headers)\n",
    "\n",
    "        if not r.ok:\n",
    "            print(f\"Error while retrieve {gene_name}: {r.status_code}\")\n",
    "            continue\n",
    "        \n",
    "        data = r.json() # Store extract data in json format\n",
    "        sequence = data.get('seq', '') # extract the sequence from data in json format\n",
    "        #print(len(sequence))\n",
    "        trimmedsequence = sequence[:3000]\n",
    "        #print(len(trimmedsequence))\n",
    "        # Write the fasta file row\n",
    "        fasta_file.write(f'>{gene_name}_{gene_id}\\n{trimmedsequence}\\n')\n",
    "        \n",
    "        # Print to check\n",
    "        print(f\">{gene_name}_{gene_id}\\n{trimmedsequence}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efe81c3a",
   "metadata": {},
   "source": [
    "## Generate the csv files for matches (motif)\n",
    "#### Use to find the exact match strategy in promoter regioin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "63bedbe4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IFNB1\n",
      "0\n",
      "\n",
      "IFIT1\n",
      "0\n",
      "\n",
      "MX1\n",
      "0\n",
      "\n",
      "TLR7\n",
      "0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import re\n",
    "import csv\n",
    "\n",
    "# Set the motif pattern\n",
    "motif_sequence = \"GAAACCGAAA\"\n",
    "reverse_motif_sequence = str(Seq(motif_sequence).reverse_complement())\n",
    "forward_pattern = re.compile(r\"GAAACCGAAA\")\n",
    "reverse_pattern = re.compile(r'TTTCGGTTTC')\n",
    "\n",
    "# Empty list to store the match results\n",
    "results = []\n",
    "\n",
    "# Parse the promoter_sequences fasta file\n",
    "# Extract the gene information, sequence then store at variable\n",
    "for record in SeqIO.parse(\"Promoter_sequences.fasta\", \"fasta\"):\n",
    "    gene_info = record.id\n",
    "    sequence = str(record.seq)\n",
    "    rev_sequence = str(Seq(sequence).reverse_complement())\n",
    "\n",
    "    # Split the gene info into gene_name and gene_id\n",
    "    if \"_\" in gene_info:\n",
    "        gene_name, gene_id = gene_info.split(\"_\", 1)\n",
    "    else:\n",
    "        gene_name, gene_id = gene_info, \"NA\"\n",
    "\n",
    "    # Find the matches in forward strand\n",
    "    forward_match = list(forward_pattern.finditer(sequence)) # make a list to count the match numbers\n",
    "    forward_position = [] # prepare the empty list to add match positions\n",
    "    for match in forward_match:\n",
    "        start = match.start()\n",
    "        end = match.end()\n",
    "        forward_position.append(f'+[{start}-{end}]')\n",
    "\n",
    "    # Find the matches in reverse strand\n",
    "    reverse_match = list(reverse_pattern.finditer(rev_sequence)) # make a list to count the match numbers\n",
    "    reverse_position = [] # prepare the empty list to add match positions\n",
    "    for match in reverse_match:\n",
    "        start = match.start()\n",
    "        end = match.end()\n",
    "        reverse_position.append(f'-[{start}-{end}]')\n",
    "\n",
    "    # Count total hits\n",
    "    total_hits = len(forward_match) + len(reverse_match)\n",
    "    all_positions = forward_position + reverse_position\n",
    "    all_positions = \" | \".join(all_positions)\n",
    "    \n",
    "    print(gene_name)\n",
    "    print(total_hits)\n",
    "    print(all_positions)\n",
    "\n",
    "    # Add colum data into results list\n",
    "    if total_hits > 0:\n",
    "        results.append([\n",
    "            gene_name, \n",
    "            gene_id, \n",
    "            motif_sequence, \n",
    "            total_hits, \n",
    "            all_positions, \n",
    "            sequence])\n",
    "        \n",
    "# Generate the csv file for hits\n",
    "with open(\"motif_scan_results.csv\", \"w\", newline=\"\") as csvfile:\n",
    "    writer = csv.writer(csvfile)\n",
    "    writer.writerow([\n",
    "        \"Gene_name\",\n",
    "        \"Gene_ID\",\n",
    "        \"Motif_Sequence\",\n",
    "        \"Number_of_Hits\",\n",
    "        \"Position_of_Hits\",\n",
    "        \"Promoter_Sequence\"\n",
    "    ])\n",
    "    writer.writerows(results)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86535333",
   "metadata": {},
   "source": [
    "## Search with regex\n",
    "#### Not exactly match"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "5f3cb2f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IFNB1\n",
      "0\n",
      "\n",
      "IFIT1\n",
      "2\n",
      "+[2697-2707] | -[293-303]\n",
      "MX1\n",
      "2\n",
      "+[1920-1930] | -[1070-1080]\n",
      "TLR7\n",
      "0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import re\n",
    "import csv\n",
    "\n",
    "# Set the motif pattern\n",
    "motif_sequence = \"GAAACCGAAA\"\n",
    "reverse_motif_sequence = str(Seq(motif_sequence).reverse_complement())\n",
    "forward_pattern = re.compile(r\"GAAA[ATGC]{2}GAAA\")\n",
    "reverse_pattern = re.compile(r\"TTTC[ATGC]{2}TTTC\")\n",
    "\n",
    "# Empty list to store the match results\n",
    "results = []\n",
    "\n",
    "# Parse the promoter_sequences fasta file\n",
    "# Extract the gene information, sequence then store at variable\n",
    "for record in SeqIO.parse(\"Promoter_sequences.fasta\", \"fasta\"):\n",
    "    gene_info = record.id\n",
    "    sequence = str(record.seq)\n",
    "    rev_sequence = str(Seq(sequence).reverse_complement())\n",
    "\n",
    "    # Split the gene info into gene_name and gene_id\n",
    "    if \"_\" in gene_info:\n",
    "        gene_name, gene_id = gene_info.split(\"_\", 1)\n",
    "    else:\n",
    "        gene_name, gene_id = gene_info, \"NA\"\n",
    "\n",
    "    # Find the matches in forward strand\n",
    "    forward_match = list(forward_pattern.finditer(sequence)) # make a list to count the match numbers\n",
    "    forward_position = [] # prepare the empty list to add match positions\n",
    "    for match in forward_match:\n",
    "        start = match.start()\n",
    "        end = match.end()\n",
    "        forward_position.append(f'+[{start}-{end}]')\n",
    "\n",
    "    # Find the matches in reverse strand\n",
    "    reverse_match = list(reverse_pattern.finditer(rev_sequence)) # make a list to count the match numbers\n",
    "    reverse_position = [] # prepare the empty list to add match positions\n",
    "    for match in reverse_match:\n",
    "        start = match.start()\n",
    "        end = match.end()\n",
    "        reverse_position.append(f'-[{start}-{end}]')\n",
    "\n",
    "    # Count total hits\n",
    "    total_hits = len(forward_match) + len(reverse_match)\n",
    "    all_positions = forward_position + reverse_position\n",
    "    all_positions = \" | \".join(all_positions)\n",
    "    \n",
    "    print(gene_name)\n",
    "    print(total_hits)\n",
    "    print(all_positions)\n",
    "\n",
    "    # Add colum data into results list\n",
    "    if total_hits > 0:\n",
    "        results.append([\n",
    "            gene_name, \n",
    "            gene_id, \n",
    "            motif_sequence, \n",
    "            total_hits, \n",
    "            all_positions, \n",
    "            sequence])\n",
    "        \n",
    "# Generate the csv file for hits\n",
    "with open(\"regex_motif_scan_results.csv\", \"w\", newline=\"\") as csvfile:\n",
    "    writer = csv.writer(csvfile)\n",
    "    writer.writerow([\n",
    "        \"Gene_name\",\n",
    "        \"Gene_ID\",\n",
    "        \"Motif_Sequence\",\n",
    "        \"Number_of_Hits\",\n",
    "        \"Position_of_Hits\",\n",
    "        \"Promoter_Sequence\"\n",
    "    ])\n",
    "    writer.writerows(results)\n",
    "    \n",
    "    \n",
    "\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd8d46f5",
   "metadata": {},
   "source": [
    "## Find the matches in forward strand\n",
    "#### Using two type of pattern - regex\n",
    "#### forward pattern 'GAAANNGAAA'\n",
    "#### reverse pattern 'TTTCNNTTTC'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52a54793",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[['IFIT1', 'ENSG00000185745', 'forward: GAAANNGAAA / reverse: TTTCNNTTTC', 4, '+[2697-2707] | -[1008-1018] | -[2961-2971] | -[2974-2984]', 'GCTGATAAGAGATTTTAGACTACAAGAGATGACCCTGGAAAAGACAGGGGAAATTTCTATGGAGCCCATTCCTTAATAAGTTGTGATTGGATATTTGAGACCCTTCAGCAGAAATGGCTTAGCATTTGTCCATATTATCATATTATAATTTTTAAAAAATACACTCAGAATAACATTCTTTTTCCATATCATCCTGTGAATGAAGAAGAGACTAAGGAATTCATTACCCATTGTGTAGGTGAGGAAACAAGGCATTGAGTCGTAAAATACTGAATCTGGGCCTCACAGAACAGGAAGTGGGTCCAGGGATTGAGTCCCATATCTCTCAGTCAGAGAATTTCAGACATACAATAGACATTAGGGATAATTGAGGCCCTGGTTCCCGGCTCAGGGCGCAGGTCTACATCAAAATCACCTGAAGATTTCTTTCTTTTTTTTTTTTTTTTTTTTTTTGAGATGGAGTTTCACTCTTGTTGCCCAGGCTGGAGTGCAATGAAATGGCACGATCTCGGCTCACCACAACCTCCACCTCCCGGGTTCAAGCGATTCTCCTGCCTCAGCCTCCCGAGTAGCTGGGATTACAGGCATGTGCCACCACGCCCGGCTAATTATGTATTTTTAGTAGAGACGGGGTTTCTCCATGTTAGTCAGGCTGGTCTCGAACTCCCAACCTCAGGTGATCCGCCCGTCTTGTCCTCCCAAAATGCTGGGATTACAGGCGTGAGCCACTGTGCCCAGGCTGAAGATTTCTTTTTAATACTGAGCCAGAGGCCTGTACCTAAGTATATGTTCAATTGTTCTGGAATGGGCATAGTTAGCATTATTTTTGAGATTTACCTGGCTGATTCTAATGTGCAGTGAGGGTTGAGATCTGCCAACCTATTCCAAATGACTTCTCAGCTGAGGAAATTGAGATTCATAGAGGTGAAGCAGTTTATCATAACTGGTCAGTCCCTTGCAGAGCCCAGCTTAGACTCCTGGGAATGAACTCACAACCTAGTGCTGAAATTTCATTTTCTAAATGTTTGATCCTCTCTGGTTCCCATTTTCTTTTTTATCTGTAAACACACACACACACGCGCGCGATTGGCTTTTCTTTCCTTAATGCTGTTCTATCCTGTGTCCCTCATGGGTTTGTTTAGCCCAAAGAGACACTAAGCAAGATTCTGAATACCTAATCAGCCTGGGCCATAGCCCCGTAGAACCAAGACTGCTGCCAAATTCACAGCAATGGCAAGATAAACTTTTTGTTCATATCTCTGCATGCTATCTAGTGCATTGGCAAATTGAAAACCCATTTTTTCCCATAAGAATTCCAAAGACATCTTGCTCCCCAAAAGATGTTTTCTAATTCATCTGTTTCAGCATTTGCTTATGGCCTGATCATGTATCTGAGTGCAAGACAAAACCTGTCCTATTAATTCAGAAATTCTTATTTGTGTAGAACTTTCTAACTAGTTGTTGAACTCTCTGTGGAAGCTTACAGAATTGGTGAGAGCTGACATGGGAGATGAGTCTGGCAGGGAGTTGAGCAGAAACTGTTTAGATCTGGCAGTGTGCCCAGTTAAGTGTTTCCAAAAGGAAAATACTCTTGCACCTGCTCCTCTTCTGCTTCTTATGTTAAAAGACAAAGTAGGTAGGTTTTTTTAAAGGAAACATTGTTTTTATATATTGTTATGATTATTTGAGACAGGGGCTCACTCTGTTGCCCAGGCTGGAGTGCAGTGGTGTGACCATGGCTCACTGCAGCCTCTACCTCCTGGGCTCAAGTGAGCCTCCCACCTCAGCCTCCTGAGTAGCTGAGACCACAGACACACACCACCACACCCAGCTTTTTTTTTCAGTTTTTTTGTAGAGTTGGGGTCTCCCTATGTTGTCCAGGCTGGTCTACAATTCCTGGGCTCAAGTGATCCTCCCGCCTTGCCTTCCAAAATACTGGGACTTACAGGCCTGAGCCAATGCACCCGGACAACAATTTTTTATTTTTTATTTATTTATTTTTTTAGACGGAGTCTCGCTCTGTCACCAGACTGGAGTACAGTGGTGTGATCTCGGCTCACTGCAACCTCTGCCTCCCAGGTTCAAGCAATTCCCCTGCCTCAGCCTCTCGAGTAGGTGGGACTACAGGTGCACACCACCACACCCAGCTAATTTTTTGTATTTTAGTAGAGAGGGGGTTTCACCATGTTGGCCACGATGGTCTCCATCTCCTGACCTTGTCATCCGCCCACCTTGGCCTCCCAAAGTGCTGGGACTACAGGCATGAGCCACCGCACCCAGCCAAGAATCATTATTTTTAACTTGATGACTGAAAATAATAATAATAATAGTTACCACTTATTTGCATGCTTCTATGTGCCAGGTAGTTGCTAACTATTTAAACTCAAATTCCATGAACTGTAGTGGAGGTTGTACTGGAATTTGATTCAGAATGACAGTGTCCATGATGGAGCAATAGAGGGCTCTCTATTTCAAACCATACCTCCTTGCTTTTACCTCCTGCCTAAGTCATCAGGGGTTAGAAGGCTTTCTAGGTATTGGTCTCTTTCCTTCATTCCTAAACCAGATTGGTTGCTTATTTCCGTCAAGCTGAAACCAAAAGTAAGCAACCAAAAAGCAACCAGCAACCAAAAGCCTTGTTACTCAATTAATTAAGAGTAGATTTTTATATTTGATAGTAGGTTCCTTCTAAATATAGAAACTGAAAATAGAGCTATCTCCTTCAATTCTCCTTTTTCTGTGTATTCATCCAGAATCCAGCCACCAACTGCCACAATAGGCAGCAATGGACTGATGTTCTTTAGGGAGGACGTGAATCTCGTTCCAAATGCTGGCCAGTCATTGGGTTTCTGCAGCACTAGAAACATCTATGGTTGCAGGTCTGCAGTTTATCTGTTTTAAAATAGAAACAAAGTTTCATTCCCCACCCCCCCCCGTCAGCAGGAATTCCGCTAGCTTTAGTTTCACTTTCCCCTTTCGGTTTCCCTAGGTTTCCAACTT']]\n",
      "[['IFIT1', 'ENSG00000185745', 'forward: GAAANNGAAA / reverse: TTTCNNTTTC', 4, '+[2697-2707] | -[1008-1018] | -[2961-2971] | -[2974-2984]', 'GCTGATAAGAGATTTTAGACTACAAGAGATGACCCTGGAAAAGACAGGGGAAATTTCTATGGAGCCCATTCCTTAATAAGTTGTGATTGGATATTTGAGACCCTTCAGCAGAAATGGCTTAGCATTTGTCCATATTATCATATTATAATTTTTAAAAAATACACTCAGAATAACATTCTTTTTCCATATCATCCTGTGAATGAAGAAGAGACTAAGGAATTCATTACCCATTGTGTAGGTGAGGAAACAAGGCATTGAGTCGTAAAATACTGAATCTGGGCCTCACAGAACAGGAAGTGGGTCCAGGGATTGAGTCCCATATCTCTCAGTCAGAGAATTTCAGACATACAATAGACATTAGGGATAATTGAGGCCCTGGTTCCCGGCTCAGGGCGCAGGTCTACATCAAAATCACCTGAAGATTTCTTTCTTTTTTTTTTTTTTTTTTTTTTTGAGATGGAGTTTCACTCTTGTTGCCCAGGCTGGAGTGCAATGAAATGGCACGATCTCGGCTCACCACAACCTCCACCTCCCGGGTTCAAGCGATTCTCCTGCCTCAGCCTCCCGAGTAGCTGGGATTACAGGCATGTGCCACCACGCCCGGCTAATTATGTATTTTTAGTAGAGACGGGGTTTCTCCATGTTAGTCAGGCTGGTCTCGAACTCCCAACCTCAGGTGATCCGCCCGTCTTGTCCTCCCAAAATGCTGGGATTACAGGCGTGAGCCACTGTGCCCAGGCTGAAGATTTCTTTTTAATACTGAGCCAGAGGCCTGTACCTAAGTATATGTTCAATTGTTCTGGAATGGGCATAGTTAGCATTATTTTTGAGATTTACCTGGCTGATTCTAATGTGCAGTGAGGGTTGAGATCTGCCAACCTATTCCAAATGACTTCTCAGCTGAGGAAATTGAGATTCATAGAGGTGAAGCAGTTTATCATAACTGGTCAGTCCCTTGCAGAGCCCAGCTTAGACTCCTGGGAATGAACTCACAACCTAGTGCTGAAATTTCATTTTCTAAATGTTTGATCCTCTCTGGTTCCCATTTTCTTTTTTATCTGTAAACACACACACACACGCGCGCGATTGGCTTTTCTTTCCTTAATGCTGTTCTATCCTGTGTCCCTCATGGGTTTGTTTAGCCCAAAGAGACACTAAGCAAGATTCTGAATACCTAATCAGCCTGGGCCATAGCCCCGTAGAACCAAGACTGCTGCCAAATTCACAGCAATGGCAAGATAAACTTTTTGTTCATATCTCTGCATGCTATCTAGTGCATTGGCAAATTGAAAACCCATTTTTTCCCATAAGAATTCCAAAGACATCTTGCTCCCCAAAAGATGTTTTCTAATTCATCTGTTTCAGCATTTGCTTATGGCCTGATCATGTATCTGAGTGCAAGACAAAACCTGTCCTATTAATTCAGAAATTCTTATTTGTGTAGAACTTTCTAACTAGTTGTTGAACTCTCTGTGGAAGCTTACAGAATTGGTGAGAGCTGACATGGGAGATGAGTCTGGCAGGGAGTTGAGCAGAAACTGTTTAGATCTGGCAGTGTGCCCAGTTAAGTGTTTCCAAAAGGAAAATACTCTTGCACCTGCTCCTCTTCTGCTTCTTATGTTAAAAGACAAAGTAGGTAGGTTTTTTTAAAGGAAACATTGTTTTTATATATTGTTATGATTATTTGAGACAGGGGCTCACTCTGTTGCCCAGGCTGGAGTGCAGTGGTGTGACCATGGCTCACTGCAGCCTCTACCTCCTGGGCTCAAGTGAGCCTCCCACCTCAGCCTCCTGAGTAGCTGAGACCACAGACACACACCACCACACCCAGCTTTTTTTTTCAGTTTTTTTGTAGAGTTGGGGTCTCCCTATGTTGTCCAGGCTGGTCTACAATTCCTGGGCTCAAGTGATCCTCCCGCCTTGCCTTCCAAAATACTGGGACTTACAGGCCTGAGCCAATGCACCCGGACAACAATTTTTTATTTTTTATTTATTTATTTTTTTAGACGGAGTCTCGCTCTGTCACCAGACTGGAGTACAGTGGTGTGATCTCGGCTCACTGCAACCTCTGCCTCCCAGGTTCAAGCAATTCCCCTGCCTCAGCCTCTCGAGTAGGTGGGACTACAGGTGCACACCACCACACCCAGCTAATTTTTTGTATTTTAGTAGAGAGGGGGTTTCACCATGTTGGCCACGATGGTCTCCATCTCCTGACCTTGTCATCCGCCCACCTTGGCCTCCCAAAGTGCTGGGACTACAGGCATGAGCCACCGCACCCAGCCAAGAATCATTATTTTTAACTTGATGACTGAAAATAATAATAATAATAGTTACCACTTATTTGCATGCTTCTATGTGCCAGGTAGTTGCTAACTATTTAAACTCAAATTCCATGAACTGTAGTGGAGGTTGTACTGGAATTTGATTCAGAATGACAGTGTCCATGATGGAGCAATAGAGGGCTCTCTATTTCAAACCATACCTCCTTGCTTTTACCTCCTGCCTAAGTCATCAGGGGTTAGAAGGCTTTCTAGGTATTGGTCTCTTTCCTTCATTCCTAAACCAGATTGGTTGCTTATTTCCGTCAAGCTGAAACCAAAAGTAAGCAACCAAAAAGCAACCAGCAACCAAAAGCCTTGTTACTCAATTAATTAAGAGTAGATTTTTATATTTGATAGTAGGTTCCTTCTAAATATAGAAACTGAAAATAGAGCTATCTCCTTCAATTCTCCTTTTTCTGTGTATTCATCCAGAATCCAGCCACCAACTGCCACAATAGGCAGCAATGGACTGATGTTCTTTAGGGAGGACGTGAATCTCGTTCCAAATGCTGGCCAGTCATTGGGTTTCTGCAGCACTAGAAACATCTATGGTTGCAGGTCTGCAGTTTATCTGTTTTAAAATAGAAACAAAGTTTCATTCCCCACCCCCCCCCGTCAGCAGGAATTCCGCTAGCTTTAGTTTCACTTTCCCCTTTCGGTTTCCCTAGGTTTCCAACTT'], ['MX1', 'ENSG00000157601', 'forward: GAAANNGAAA / reverse: TTTCNNTTTC', 1, '+[1920-1930]', 'AATGAAATAATGGCATTCACAGCAACCTGGATGGAGTTGGAGACCATTATTCTAAGTGAAGTAACTCGGGAATAAAAACCAAACATTGTATATTCTCACTCATAAGTGAGATTATACAAGCTATGAGGATGCAAAGTCATAAGAATGATACAATAAACATTGGGAACTCAGGGGGAAGGGTGAGGGTGGGGTGAGGGATAAAAGACTAGACATCGGGTACAGTATACACTGCTTGGGTGATGGGTGCACCAAAATCCCAGAAATCACGATTAAAGAACTTGTCAATGTAACCAAACACCACCTGTTCCTCAAAAACTATTGAAATAAAAAATAAATACATAAAAGAAGTTGAGGTTAAATTAATCTAAGTAGGGAGTGCATTTGGGCCAAGCTTGAGGATTGCAACGTTGGAGCACAGATTTAAGTTGCCCTGAATCTACACTCTGATAAGTGGCATTTATAAGTGGATTTTTAAAGGCAAAAGGGGGAGACAGGGAGAGGGCTGATACAAAGCTGTTTGTCAGGAATTCTTATTGGTTTACAGAAATAACATTGATAAGTGCTTATTAGTCCATTTTCATGCTGCTGATAAAGACATACCTCAGACTGGGCAATTTACAAAAGAAAGAGGTTTATTGGACCTATAGTTCCACATGGCTGAGGAGGCCTCACAATCATGGTGGAAGGCAAGGAGGAGCAAGTCACATCTTACATGGATGACAGCAGGAAAAAAGAGAGAGCTTGTGCAGAGAAACTCCCTTTTTTAAAACCATCAGATCTTGTGAGACCCGTTCAGGATCATGAGAACAGCACAGGAAAGACCCACCCCCAAGATTCAATCATCTCCCACCGGGTCTCTCCCACAACATGAGTGAATTATGGGAGCTACAGGATGAGATTTGGGTGAGGACACAGAGCCAAACCATATCAGAGTGATTGACTATACATTGTTAAGCTACAGGGTGTGGATTACAGCGTCCAGTGTGGCATTATTAGGTTAATCCATAGCTATTAATACCTGTGGCAATAGTAAGAGGTTTCAAGAGATGAACACACAGCTGAAAGTGGGGAGTAGGGAGTGGTTGCTGTCTTATTTTAATGTCTCTCTGGGCCTGATAATTTTAAAAGCCTTGCATTCTTCGGATACAAGTTCCTTTCTTTTTTTCAAGAACAACAGAGAAAGGCTGCCTGGCTACCTTTACCTGCAGCACCTCTGATAGTGGCTGCTTGGCAGCTGGTTAGGTAAACCCAATTTCTACTGGATAAACTTCATGGACATTCACCCATGTTAGTACTGGGGAATGCTCTGAAATAGCAAACTTGTCCATACAGCTCCCTTGCTGCTCAGAGGGTCTGGAGAGGAACAGCAGAGGCCCAGCCCCAAATACTTCTTGTACCGGCTTCTATTATGAGAGGTGATGGCACCTTTAGGCACTTATATTTCAGGTAGAAGGAGCAGGAATAAAGAAGAGAGGGCCACAATCTGAAATATCAGTTCTGGATCTCCCTGGGGGCAAGACACTGGCCCTTAGCTGAGCTATTCTGTCATCACTGTGAGCCCTCATGCTCGACCTCACCGGAGCACCCCTCGATCATGTGCTGAATGCCTAGCACGCGCCTGGCACTGTGGGGTACAGAAATCTAGAGCACAGAGCTGAGTTTTCAGGGCTTAGCTGTCCTCACATCTGTGACACAAACAAGGATCAAGGAAAGTTTTGTTTTTGTTATTTTTATTTTTTTTTAAGTTGGGGCCTTACTCTGTAGCCCAAGCTAGAGTGCAGTGGTGCAAACATAGCTCATTGCACCCTCGAACTCCTGGGCTCAAGTCATCCTCCCACCTTGGCCTCCAGAGTAGCTGGAAGCCCAGCTCAAGGAAAGCATTAACATGTTCTACTTTTCACCATAACTCCTGTGTATTGGGAAATGGAAAATGAGCTTTTGAAGTGGCCTATTCGCTTCCTCCATGAGTACTTCCTGAACACTTTCCTCTTGGCCACTGCTCCCAAGACCCCCTCAAAGTCTCTCTCCTGGACTTCCAGCTTCTGGAATGAGCCCCAAGAGTGGCACTTAGCTCCAGCTCCTAGTAGCTCACTCCTTCCTTGCTCACCTGTGCCCCCCACCTGTCTCCTTAGGGGCAGTGGCTTTGTCTTACTCATTTGCATATCACAAATGTCCAGCATAGAAGCCAGCATGGGGCCCAGTAAGTGTTTGTGGAGAAAGAAAGGGAAGAGGTGAAGAAGGGAGGAGACGGAGGGAGAAAAGGAGGGCAAGAGAGGGGATGTCATGTCATGTCATGTCATGTCATGTCAGTTCTCCCTTCCTAGATCTGGACAAGGGAAAGAAAAGCAAGCAAGCCTGACCGCCTCAGTGCAGACCCTCTAAGATGCAAGCCTTCCCCAGGTAAGACAGTTCCGCAGGGTGCTTCTGTGGGCTGGCTTTCTGAGCAGCCATCTCAAAGTATGCGAAAGAAATAGATTTTGGGCCGGGCACGGTGGCTCATGCCTGTAATCCCAGCACTTTAGGAGGCCGAGGCGTGTGGATCACCTGAAGTCAGGAGTTCAAGACCAGCCTGGCCAACATGGTGAGACCTTGTATCTACTAAAAATACAAAAACTAGCCGGGCTTGGTGGCGCATGCCTGTAGTCCCAGCTACTCAGGAGGCTGAGGCAGGAGAATCGCTTGAACCCAGGAGGTAAATGTTGCAGTAAGCCGAGATTGCCCCACTGCACTCCAGCCTGGATGACAAGAGTGAGACTCCATCTCAAAAAAAAAAAAAAAAAAAGATTTTAGGGGTAAAATATTTTGCTTTCCTTCAGTGGCTGGGCCACCTGTACACAATGCCTGAACCCAGGTCCTGCAGCCCCCAGCTTGGCTCTCCAAAGCTCACCAGTATCAAAGGATTTCAGCTTCTGCTCCTCTCCGCAGGGCTCTCTGGTGGGCCCCAGACAGAGATGAAGGAAAAGTCTGGCCTTGACCTTGAGGACCAAAAGCGACCACGTC']]\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import re\n",
    "import csv\n",
    "\n",
    "# Set the motif pattern\n",
    "motif_sequence = \"forward: GAAANNGAAA / reverse: TTTCNNTTTC\"\n",
    "forward_pattern = re.compile(r\"GAAA[ATGC]{2}GAAA\")\n",
    "reverse_pattern = re.compile(r\"TTTC[ATGC]{2}TTTC\")\n",
    "\n",
    "# Empty list to store the match results\n",
    "results = []\n",
    "\n",
    "# Parse the promoter_sequences fasta file\n",
    "# Extract the gene information, sequence then store at variable\n",
    "for record in SeqIO.parse(\"Promoter_sequences.fasta\", \"fasta\"):\n",
    "    gene_info = record.id\n",
    "    sequence = str(record.seq)\n",
    "\n",
    "    # Split the gene info into gene_name and gene_id\n",
    "    if \"_\" in gene_info:\n",
    "        gene_name, gene_id = gene_info.split(\"_\", 1)\n",
    "    else:\n",
    "        gene_name, gene_id = gene_info, \"NA\"\n",
    "\n",
    "    # Find the matches in forward strand\n",
    "    forward_match = list(forward_pattern.finditer(sequence)) # make a list to count the match numbers\n",
    "    forward_position = [] # prepare the empty list to add match positions\n",
    "    for match in forward_match:\n",
    "        start = match.start()\n",
    "        end = match.end()\n",
    "        forward_position.append(f'+[{start}-{end}]')\n",
    "\n",
    "    # Find the matches in reverse strand\n",
    "    reverse_match = list(reverse_pattern.finditer(sequence)) # make a list to count the match numbers\n",
    "    reverse_position = [] # prepare the empty list to add match positions\n",
    "    for match in reverse_match:\n",
    "        start = match.start()\n",
    "        end = match.end()\n",
    "        reverse_position.append(f'-[{start}-{end}]')\n",
    "\n",
    "    # Count total hits\n",
    "    total_hits = len(forward_match) + len(reverse_match)\n",
    "    all_positions = forward_position + reverse_position\n",
    "    all_positions = \" | \".join(all_positions)\n",
    "\n",
    "    # Add colum data into results list\n",
    "    if total_hits > 0:\n",
    "        results.append([\n",
    "            gene_name, \n",
    "            gene_id, \n",
    "            motif_sequence, \n",
    "            total_hits, \n",
    "            all_positions, \n",
    "            sequence])\n",
    "\n",
    "# Generate the csv file for hits\n",
    "with open(\"regex_motif_scan_forward_results.csv\", \"w\", newline=\"\") as csvfile:\n",
    "    writer = csv.writer(csvfile)\n",
    "    writer.writerow([\n",
    "        \"Gene_name\",\n",
    "        \"Gene_ID\",\n",
    "        \"Motif_Sequence\",\n",
    "        \"Number_of_Hits\",\n",
    "        \"Position_of_Hits\",\n",
    "        \"Promoter_Sequence\"\n",
    "    ])\n",
    "    writer.writerows(results)\n",
    "    \n",
    "\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03c71a25",
   "metadata": {},
   "source": [
    "## Find the matches in forward strand\n",
    "#### Using two type of pattern - exact match\n",
    "#### forward pattern 'GAAACCGAAA'\n",
    "#### reverse pattern 'TTTCGGTTTC'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1d9d2f63",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'Promoter_sequences.fasta'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[3], line 16\u001b[0m\n\u001b[1;32m     12\u001b[0m results \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m     14\u001b[0m \u001b[38;5;66;03m# Parse the promoter_sequences fasta file\u001b[39;00m\n\u001b[1;32m     15\u001b[0m \u001b[38;5;66;03m# Extract the gene information, sequence then store at variable\u001b[39;00m\n\u001b[0;32m---> 16\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m record \u001b[38;5;129;01min\u001b[39;00m \u001b[43mSeqIO\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mparse\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mPromoter_sequences.fasta\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mfasta\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m:\n\u001b[1;32m     17\u001b[0m     gene_info \u001b[38;5;241m=\u001b[39m record\u001b[38;5;241m.\u001b[39mid\n\u001b[1;32m     18\u001b[0m     sequence \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mstr\u001b[39m(record\u001b[38;5;241m.\u001b[39mseq)\n",
      "File \u001b[0;32m~/Library/Python/3.9/lib/python/site-packages/Bio/SeqIO/__init__.py:626\u001b[0m, in \u001b[0;36mparse\u001b[0;34m(handle, format, alphabet)\u001b[0m\n\u001b[1;32m    624\u001b[0m iterator_generator \u001b[38;5;241m=\u001b[39m _FormatToIterator\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;28mformat\u001b[39m)\n\u001b[1;32m    625\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m iterator_generator:\n\u001b[0;32m--> 626\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43miterator_generator\u001b[49m\u001b[43m(\u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    627\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mformat\u001b[39m \u001b[38;5;129;01min\u001b[39;00m AlignIO\u001b[38;5;241m.\u001b[39m_FormatToIterator:\n\u001b[1;32m    628\u001b[0m     \u001b[38;5;66;03m# Use Bio.AlignIO to read in the alignments\u001b[39;00m\n\u001b[1;32m    629\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m (r \u001b[38;5;28;01mfor\u001b[39;00m alignment \u001b[38;5;129;01min\u001b[39;00m AlignIO\u001b[38;5;241m.\u001b[39mparse(handle, \u001b[38;5;28mformat\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m r \u001b[38;5;129;01min\u001b[39;00m alignment)\n",
      "File \u001b[0;32m~/Library/Python/3.9/lib/python/site-packages/Bio/SeqIO/FastaIO.py:196\u001b[0m, in \u001b[0;36mFastaIterator.__init__\u001b[0;34m(self, source, alphabet)\u001b[0m\n\u001b[1;32m    194\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m alphabet \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    195\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mThe alphabet argument is no longer supported\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m--> 196\u001b[0m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;21;43m__init__\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43msource\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfmt\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mFasta\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    197\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    198\u001b[0m     line \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mnext\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstream)\n",
      "File \u001b[0;32m~/Library/Python/3.9/lib/python/site-packages/Bio/SeqIO/Interfaces.py:81\u001b[0m, in \u001b[0;36mSequenceIterator.__init__\u001b[0;34m(self, source, alphabet, fmt)\u001b[0m\n\u001b[1;32m     79\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(source, _PathLikeTypes):\n\u001b[1;32m     80\u001b[0m     mode \u001b[38;5;241m=\u001b[39m modes[\u001b[38;5;241m0\u001b[39m]\n\u001b[0;32m---> 81\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstream \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43msource\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mr\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     82\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m     83\u001b[0m     value \u001b[38;5;241m=\u001b[39m source\u001b[38;5;241m.\u001b[39mread(\u001b[38;5;241m0\u001b[39m)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'Promoter_sequences.fasta'"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import re\n",
    "import csv\n",
    "\n",
    "# Set the motif pattern\n",
    "motif_sequence = \"forward: GAAACCGAAA / reverse: TTTCGGTTTC\"\n",
    "forward_pattern = re.compile(r\"GAAACCGAAA\")\n",
    "reverse_pattern = re.compile(r\"TTTCGGTTTC\")\n",
    "\n",
    "# Empty list to store the match results\n",
    "results = []\n",
    "\n",
    "# Parse the promoter_sequences fasta file\n",
    "# Extract the gene information, sequence then store at variable\n",
    "for record in SeqIO.parse(\"Promoter_sequences.fasta\", \"fasta\"):\n",
    "    gene_info = record.id\n",
    "    sequence = str(record.seq)\n",
    "\n",
    "    # Split the gene info into gene_name and gene_id\n",
    "    if \"_\" in gene_info:\n",
    "        gene_name, gene_id = gene_info.split(\"_\", 1)\n",
    "    else:\n",
    "        gene_name, gene_id = gene_info, \"NA\"\n",
    "\n",
    "    # Find the matches in forward strand\n",
    "    forward_match = list(forward_pattern.finditer(sequence)) # make a list to count the match numbers\n",
    "    forward_position = [] # prepare the empty list to add match positions\n",
    "    for match in forward_match:\n",
    "        start = match.start()\n",
    "        end = match.end()\n",
    "        forward_position.append(f'+[{start}-{end}]')\n",
    "\n",
    "    # Find the matches in reverse strand\n",
    "    reverse_match = list(reverse_pattern.finditer(sequence)) # make a list to count the match numbers\n",
    "    reverse_position = [] # prepare the empty list to add match positions\n",
    "    for match in reverse_match:\n",
    "        start = match.start()\n",
    "        end = match.end()\n",
    "        reverse_position.append(f'-[{start}-{end}]')\n",
    "\n",
    "    # Count total hits\n",
    "    total_hits = len(forward_match) + len(reverse_match)\n",
    "    all_positions = forward_position + reverse_position\n",
    "    all_positions = \" | \".join(all_positions)\n",
    "\n",
    "    # Add colum data into results list\n",
    "    if total_hits > 0:\n",
    "        results.append([\n",
    "            gene_name, \n",
    "            gene_id, \n",
    "            motif_sequence, \n",
    "            total_hits, \n",
    "            all_positions, \n",
    "            sequence])\n",
    "\n",
    "# Generate the csv file for hits\n",
    "with open(\"motif_scan_forward_results.csv\", \"w\", newline=\"\") as csvfile:\n",
    "    writer = csv.writer(csvfile)\n",
    "    writer.writerow([\n",
    "        \"Gene_name\",\n",
    "        \"Gene_ID\",\n",
    "        \"Motif_Sequence\",\n",
    "        \"Number_of_Hits\",\n",
    "        \"Position_of_Hits\",\n",
    "        \"Promoter_Sequence\"\n",
    "    ])\n",
    "    writer.writerows(results)\n",
    "    \n",
    "\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b013a4d",
   "metadata": {},
   "source": [
    "jaspar module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfb2b4be",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Fetched motif for IRF3: MA1418.2\n",
      " Fetched motif for IRF7: MA0772.2\n",
      " Fetched motif for IRF9: MA0653.1\n",
      "3000\n",
      "3000\n",
      "17\n",
      "[(np.int64(2903), np.float32(7.022167))]\n",
      "[(np.int64(-2920), np.float32(7.022167))]\n",
      "13\n",
      "[(np.int64(2888), np.float32(8.980127))]\n",
      "[(np.int64(-2901), np.float32(8.980127))]\n",
      "15\n",
      "[]\n",
      "[]\n",
      "3000\n",
      "3000\n",
      "17\n",
      "[(np.int64(-2827), np.float32(7.4404163)), (np.int64(2691), np.float32(9.355081)), (np.int64(-40), np.float32(13.019573)), (np.int64(-33), np.float32(7.6814137)), (np.int64(-27), np.float32(7.9259024))]\n",
      "[(np.int64(10), np.float32(7.9259024)), (np.int64(16), np.float32(7.6814137)), (np.int64(23), np.float32(13.019573)), (np.int64(-2708), np.float32(9.355081)), (np.int64(2810), np.float32(7.4404163))]\n",
      "13\n",
      "[(np.int64(-1994), np.float32(14.6928)), (np.int64(1574), np.float32(8.584064)), (np.int64(-1164), np.float32(8.768341)), (np.int64(2590), np.float32(10.45305)), (np.int64(2696), np.float32(15.829915)), (np.int64(-41), np.float32(16.261517)), (np.int64(-28), np.float32(10.871973))]\n",
      "[(np.int64(15), np.float32(10.871973)), (np.int64(28), np.float32(16.261517)), (np.int64(-2709), np.float32(15.829915)), (np.int64(-2603), np.float32(10.45305)), (np.int64(1151), np.float32(8.768341)), (np.int64(-1587), np.float32(8.584064)), (np.int64(1981), np.float32(14.6928))]\n",
      "15\n",
      "[(np.int64(2694), np.float32(8.752556))]\n",
      "[(np.int64(-2709), np.float32(8.752556))]\n",
      "3000\n",
      "3000\n",
      "17\n",
      "[]\n",
      "[]\n",
      "13\n",
      "[(np.int64(-1289), np.float32(10.570211)), (np.int64(-202), np.float32(7.902282))]\n",
      "[(np.int64(189), np.float32(7.902282)), (np.int64(1276), np.float32(10.570211))]\n",
      "15\n",
      "[]\n",
      "[]\n",
      "3000\n",
      "3000\n",
      "17\n",
      "[]\n",
      "[]\n",
      "13\n",
      "[(np.int64(2738), np.float32(7.658199)), (np.int64(2777), np.float32(9.225679))]\n",
      "[(np.int64(-2790), np.float32(9.225679)), (np.int64(-2751), np.float32(7.658199))]\n",
      "15\n",
      "[]\n",
      "[]\n",
      "Results written to motif_scan_results.csv\n"
     ]
    }
   ],
   "source": [
    "# Import the packages\n",
    "\n",
    "from pyjaspar import jaspardb\n",
    "from Bio import motifs\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import csv\n",
    "\n",
    "# Fetch motifs from JASPAR database and return as Bio.motif object\n",
    "def fetch_motifs_from_jaspar(tf_list, release=\"JASPAR2024\"):\n",
    "\n",
    "    jdb = jaspardb(release=release) # set the JASPAR object\n",
    "    motif_dict = {} # create dict for storing motif\n",
    "\n",
    "    for tf in tf_list: \n",
    "        motif_list = jdb.fetch_motifs_by_name(tf) # Fetch motifs from JASPAR by name\n",
    "        if motif_list:\n",
    "            jaspar_motif = motif_list[0] \n",
    "            motif_dict[tf] = jaspar_motif # Add to dict\n",
    "            print(f\" Fetched motif for {tf}: {jaspar_motif.matrix_id}\") # Check the TF and ID\n",
    "        else:\n",
    "            print(f\"No motif found for {tf}\")\n",
    "\n",
    "    return motif_dict\n",
    "\n",
    "# Scan promoters (fasta file) for matches and include match\n",
    "def scan_promoters_for_motifs(fasta_file, motif_dict, threshold=7.0):\n",
    "\n",
    "    results = [] # Set the empty list to store match motif's information\n",
    "\n",
    "    for record in SeqIO.parse(fasta_file, \"fasta\"): # Parse the fasta file\n",
    "        gene_info = record.id # Extract the gene_info from fasta file\n",
    "        promoter_seq = str(record.seq) # Extract the sequence data from fasta file\n",
    "        reverse_seq = str(Seq(promoter_seq).reverse_complement()) # Reverse the sequence\n",
    "        #print(len(promoter_seq))\n",
    "        #print(len(reverse_seq))\n",
    "\n",
    "        if \"_\" in gene_info:\n",
    "            gene_name, gene_id = gene_info.split(\"_\", 1)\n",
    "        else:\n",
    "            gene_name, gene_id = gene_info, \"NA\"\n",
    "\n",
    "        for tf, motif in motif_dict.items():\n",
    "            pssm = motif.pssm # Position-Specific Scoring Matrix\n",
    "            motif_len = len(motif) # check the length of motif\n",
    "            #print(motif_len)\n",
    "\n",
    "            # Find the matches in forward strand\n",
    "            forward_matches = list(pssm.search(promoter_seq, threshold=threshold))\n",
    "            print(forward_matches)\n",
    "            for pos, score in forward_matches:\n",
    "                start = pos \n",
    "                end = pos + motif_len\n",
    "                results.append([gene_name, gene_id, tf, \"+\", start, end]) # Add matched TF information to list\n",
    "\n",
    "            # Find the matches in reverse strand\n",
    "            reverse_matches = list(pssm.search(reverse_seq, threshold=threshold))\n",
    "            print(reverse_matches)\n",
    "            for pos, score in reverse_matches:\n",
    "                start = len(promoter_seq) - pos - motif_len\n",
    "                end = start + motif_len\n",
    "                results.append([gene_name, gene_id, tf, \"-\", start, end]) # Add matched TF information to list\n",
    "\n",
    "    return results\n",
    "\n",
    "# Write results to CSV with start and end columns\n",
    "def write_results_to_csv(results, filename=\"motif_scan_results.csv\"):\n",
    "    with open(filename, \"w\", newline=\"\") as csvfile:\n",
    "        writer = csv.writer(csvfile)\n",
    "        writer.writerow([\n",
    "            \"Gene_Name\", \"Gene_ID\", \"TF\", \"Strand\", \"Start\", \"End\"\n",
    "        ]) # Assign the column name\n",
    "        writer.writerows(results) # Write the row in csv file\n",
    "    print(f\"Results written to {filename}\")\n",
    "\n",
    "# Run functions\n",
    "if __name__ == \"__main__\":\n",
    "    tf_names = [\"IRF3\", \"IRF7\", \"IRF9\"] # Set the list to fetch multile transcription factors\n",
    "    fasta_file = \"Promoter_sequences.fasta\" # Sequence data to scan\n",
    "\n",
    "    motifs_dict = fetch_motifs_from_jaspar(tf_names) # Store function result in variable\n",
    "    scan_results = scan_promoters_for_motifs(fasta_file, motifs_dict, threshold=7.0) # Store function result in variable\n",
    "    write_results_to_csv(scan_results)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4aa967ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Fetched motif for IRF3: MA1418.2\n",
      "3000\n",
      "3000\n",
      "3000\n",
      "3000\n",
      "3000\n",
      "3000\n",
      "3000\n",
      "3000\n",
      "Results written to motif_scan_IRF3.csv\n"
     ]
    }
   ],
   "source": [
    "from pyjaspar import jaspardb\n",
    "from Bio import motifs\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import csv\n",
    "\n",
    "# Fetch motifs from JASPAR database and return as Bio.motif object\n",
    "def fetch_motifs_from_jaspar(tf_names, release=\"JASPAR2024\"):\n",
    "\n",
    "    jdb = jaspardb(release=release)\n",
    "    motif = jdb.fetch_motifs_by_name(tf_names)\n",
    "\n",
    "    if motif:\n",
    "        jaspar_motif = motif[0]\n",
    "        print(f\" Fetched motif for {tf_names}: {jaspar_motif.matrix_id}\")\n",
    "        return jaspar_motif\n",
    "    else:\n",
    "        print(f\"No motif found for {tf_names}\")\n",
    "        return None\n",
    "\n",
    "# Scan promoters (fasta file) for matches and include match\n",
    "def scan_promoters_for_motifs(fasta_file, motif_dict, threshold=7.0):\n",
    "\n",
    "    results = []\n",
    "\n",
    "    for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "        gene_info = record.id\n",
    "        promoter_seq = str(record.seq)\n",
    "        reverse_seq = str(Seq(promoter_seq).reverse_complement())\n",
    "        print(len(promoter_seq))\n",
    "        print(len(reverse_seq))\n",
    "\n",
    "        if \"_\" in gene_info:\n",
    "            gene_name, gene_id = gene_info.split(\"_\", 1)\n",
    "        else:\n",
    "            gene_name, gene_id = gene_info, \"NA\"\n",
    "\n",
    "            pssm = motif.pssm\n",
    "            motif_len = len(motif)\n",
    "\n",
    "            # Forward strand\n",
    "            forward_matches = list(pssm.search(promoter_seq, threshold=threshold))\n",
    "            print(forward_matches)\n",
    "            for pos, score in forward_matches:\n",
    "                start = pos\n",
    "                end = pos + motif_len\n",
    "                results.append([gene_name, gene_id, \"IRF3\", \"+\", start, end])\n",
    "\n",
    "            # Reverse strand\n",
    "            reverse_matches = list(pssm.search(reverse_seq, threshold=threshold))\n",
    "            print(reverse_matches)\n",
    "            for pos, score in reverse_matches:\n",
    "                start = len(promoter_seq) - pos - motif_len\n",
    "                if start < 0:\n",
    "                    continue # skip invalid match\n",
    "                end = start + motif_len\n",
    "                results.append([gene_name, gene_id, \"IRF3\", \"-\", start, end])\n",
    "\n",
    "    return results\n",
    "\n",
    "# Write results to CSV with start and end columns\n",
    "def write_results_to_csv(results, filename=\"motif_scan_IRF3.csv\"):\n",
    "    with open(filename, \"w\", newline=\"\") as csvfile:\n",
    "        writer = csv.writer(csvfile)\n",
    "        writer.writerow([\n",
    "            \"Gene_Name\", \"Gene_ID\", \"TF\", \"Strand\", \"Start\", \"End\"\n",
    "        ])\n",
    "        writer.writerows(results)\n",
    "    print(f\"Results written to {filename}\")\n",
    "\n",
    "# Run functions\n",
    "if __name__ == \"__main__\":\n",
    "    tf_names = \"IRF3\"\n",
    "    fasta_file = \"Promoter_sequences.fasta\"\n",
    "\n",
    "    motif = fetch_motifs_from_jaspar(tf_names)\n",
    "    scan_results = scan_promoters_for_motifs(fasta_file, motifs_dict, threshold=7.0)\n",
    "    write_results_to_csv(scan_results)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1132b23b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Fetched motif for IRF7: MA0772.2\n",
      "3000\n",
      "3000\n",
      "3000\n",
      "3000\n",
      "3000\n",
      "3000\n",
      "3000\n",
      "3000\n",
      "Results written to motif_scan_IRF7.csv\n"
     ]
    }
   ],
   "source": [
    "from pyjaspar import jaspardb\n",
    "from Bio import motifs\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import csv\n",
    "\n",
    "# Fetch motifs from JASPAR database and return as Bio.motif object\n",
    "def fetch_motifs_from_jaspar(tf_names, release=\"JASPAR2024\"):\n",
    "\n",
    "    jdb = jaspardb(release=release)\n",
    "    motif = jdb.fetch_motifs_by_name(tf_names)\n",
    "\n",
    "    if motif:\n",
    "        jaspar_motif = motif[0]\n",
    "        print(f\" Fetched motif for {tf_names}: {jaspar_motif.matrix_id}\")\n",
    "        return jaspar_motif\n",
    "    else:\n",
    "        print(f\"No motif found for {tf_names}\")\n",
    "        return None\n",
    "\n",
    "# Scan promoters (fasta file) for matches and include match\n",
    "def scan_promoters_for_motifs(fasta_file, motif_dict, threshold=7.0):\n",
    "\n",
    "    results = []\n",
    "\n",
    "    for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "        gene_info = record.id\n",
    "        promoter_seq = str(record.seq)\n",
    "        reverse_seq = str(Seq(promoter_seq).reverse_complement())\n",
    "        print(len(promoter_seq))\n",
    "        print(len(reverse_seq))\n",
    "\n",
    "        if \"_\" in gene_info:\n",
    "            gene_name, gene_id = gene_info.split(\"_\", 1)\n",
    "        else:\n",
    "            gene_name, gene_id = gene_info, \"NA\"\n",
    "\n",
    "            pssm = motif.pssm\n",
    "            motif_len = len(motif)\n",
    "\n",
    "            # Forward strand\n",
    "            forward_matches = list(pssm.search(promoter_seq, threshold=threshold))\n",
    "            print(forward_matches)\n",
    "            for pos, score in forward_matches:\n",
    "                start = pos\n",
    "                end = pos + motif_len\n",
    "                results.append([gene_name, gene_id, \"IRF7\", \"+\", start, end])\n",
    "\n",
    "            # Reverse strand\n",
    "            reverse_matches = list(pssm.search(reverse_seq, threshold=threshold))\n",
    "            print(reverse_matches)\n",
    "            for pos, score in reverse_matches:\n",
    "                start = len(promoter_seq) - pos - motif_len\n",
    "                if start < 0:\n",
    "                    continue # skip invalid match\n",
    "                end = start + motif_len\n",
    "                results.append([gene_name, gene_id, \"IRF7\", \"-\", start, end])\n",
    "\n",
    "    return results\n",
    "\n",
    "# Write results to CSV with start and end columns\n",
    "def write_results_to_csv(results, filename=\"motif_scan_IRF7.csv\"):\n",
    "    with open(filename, \"w\", newline=\"\") as csvfile:\n",
    "        writer = csv.writer(csvfile)\n",
    "        writer.writerow([\n",
    "            \"Gene_Name\", \"Gene_ID\", \"TF\", \"Strand\", \"Start\", \"End\"\n",
    "        ])\n",
    "        writer.writerows(results)\n",
    "    print(f\"Results written to {filename}\")\n",
    "\n",
    "# Run functions\n",
    "if __name__ == \"__main__\":\n",
    "    tf_names = \"IRF7\"\n",
    "    fasta_file = \"Promoter_sequences.fasta\"\n",
    "\n",
    "    motif = fetch_motifs_from_jaspar(tf_names)\n",
    "    scan_results = scan_promoters_for_motifs(fasta_file, motifs_dict, threshold=7.0)\n",
    "    write_results_to_csv(scan_results)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c69c409",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Fetched motif for IRF9: MA0653.1\n",
      "3000\n",
      "3000\n",
      "3000\n",
      "3000\n",
      "3000\n",
      "3000\n",
      "3000\n",
      "3000\n",
      "Results written to motif_scan_IRF9.csv\n"
     ]
    }
   ],
   "source": [
    "from pyjaspar import jaspardb\n",
    "from Bio import motifs\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import csv\n",
    "\n",
    "# Fetch motifs from JASPAR database and return as Bio.motif object\n",
    "def fetch_motifs_from_jaspar(tf_names, release=\"JASPAR2024\"):\n",
    "\n",
    "    jdb = jaspardb(release=release)\n",
    "    motif = jdb.fetch_motifs_by_name(tf_names)\n",
    "\n",
    "    if motif:\n",
    "        jaspar_motif = motif[0]\n",
    "        print(f\" Fetched motif for {tf_names}: {jaspar_motif.matrix_id}\")\n",
    "        return jaspar_motif\n",
    "    else:\n",
    "        print(f\"No motif found for {tf_names}\")\n",
    "        return None\n",
    "\n",
    "# Scan promoters (fasta file) for matches and include match\n",
    "def scan_promoters_for_motifs(fasta_file, motif_dict, threshold=7.0):\n",
    "\n",
    "    results = []\n",
    "\n",
    "    for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "        gene_info = record.id\n",
    "        promoter_seq = str(record.seq)\n",
    "        reverse_seq = str(Seq(promoter_seq).reverse_complement())\n",
    "        print(len(promoter_seq))\n",
    "        print(len(reverse_seq))\n",
    "\n",
    "        if \"_\" in gene_info:\n",
    "            gene_name, gene_id = gene_info.split(\"_\", 1)\n",
    "        else:\n",
    "            gene_name, gene_id = gene_info, \"NA\"\n",
    "\n",
    "            pssm = motif.pssm\n",
    "            motif_len = len(motif)\n",
    "\n",
    "            # Forward strand\n",
    "            forward_matches = list(pssm.search(promoter_seq, threshold=threshold))\n",
    "            print(forward_matches)\n",
    "            for pos, score in forward_matches:\n",
    "                start = pos\n",
    "                end = pos + motif_len\n",
    "                results.append([gene_name, gene_id, \"IRF9\", \"+\", start, end])\n",
    "\n",
    "            # Reverse strand\n",
    "            reverse_matches = list(pssm.search(reverse_seq, threshold=threshold))\n",
    "            print(reverse_matches)\n",
    "            for pos, score in reverse_matches:\n",
    "                start = len(promoter_seq) - pos - motif_len\n",
    "                if start < 0:\n",
    "                    continue # skip invalid match\n",
    "                end = start + motif_len\n",
    "                results.append([gene_name, gene_id, \"IRF9\", \"-\", start, end])\n",
    "\n",
    "    return results\n",
    "\n",
    "# Write results to CSV with start and end columns\n",
    "def write_results_to_csv(results, filename=\"motif_scan_IRF9.csv\"):\n",
    "    with open(filename, \"w\", newline=\"\") as csvfile:\n",
    "        writer = csv.writer(csvfile)\n",
    "        writer.writerow([\n",
    "            \"Gene_Name\", \"Gene_ID\", \"TF\", \"Strand\", \"Start\", \"End\"\n",
    "        ])\n",
    "        writer.writerows(results)\n",
    "    print(f\"Results written to {filename}\")\n",
    "\n",
    "# Run functions\n",
    "if __name__ == \"__main__\":\n",
    "    tf_names = \"IRF9\"\n",
    "    fasta_file = \"Promoter_sequences.fasta\"\n",
    "\n",
    "    motif = fetch_motifs_from_jaspar(tf_names)\n",
    "    scan_results = scan_promoters_for_motifs(fasta_file, motifs_dict, threshold=7.0)\n",
    "    write_results_to_csv(scan_results)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "655b6bc4",
   "metadata": {},
   "source": [
    "## Find only on forward strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "1d808d88",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched motif for IRF3: MA1418.2\n",
      "Fetched motif for IRF7: MA0772.2\n",
      "Fetched motif for IRF9: MA0653.1\n",
      "Results written to motif_scan_results_threshold_1.0.csv\n"
     ]
    }
   ],
   "source": [
    "from pyjaspar import jaspardb\n",
    "from Bio import motifs\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import csv\n",
    "\n",
    "# Fetch motifs from JASPAR\n",
    "def fetch_motifs_from_jaspar(tf_list, release=\"JASPAR2024\"):\n",
    "    jdb = jaspardb(release=release) # set the JASPAR object\n",
    "    motif_dict = {} # Create empty dict\n",
    "\n",
    "    # Iterate through transcription factors\n",
    "    for tf in tf_list:\n",
    "        motif_list = jdb.fetch_motifs_by_name(tf) \n",
    "        if motif_list:\n",
    "            jaspar_motif = motif_list[0] # Store 0 postion of motif_list to variable\n",
    "            motif_dict[tf] = jaspar_motif # Add to dict\n",
    "            print(f\"Fetched motif for {tf}: {jaspar_motif.matrix_id}\") # Print TF and TF id to confirm\n",
    "        else:\n",
    "            print(f\"No motif found for {tf}\") # Print this message when there is no motif information.\n",
    "\n",
    "    return motif_dict\n",
    "\n",
    "# Scan promoter regions using both orientations of motif on forward strand\n",
    "def scan_promoters_for_motifs(fasta_file, motif_dict, threshold=7.0):\n",
    "    results = []\n",
    "\n",
    "    # Parse the fasta file which contain promoter sequences\n",
    "    for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "        gene_info = record.id # Extract the gene information\n",
    "        promoter_seq = str(record.seq) # Extract the sequences\n",
    "\n",
    "        # Extract the gene_name and gene_id from fasta file\n",
    "        if \"_\" in gene_info:\n",
    "            gene_name, gene_id = gene_info.split(\"_\", 1)\n",
    "        else:\n",
    "            gene_name, gene_id = gene_info, \"NA\"\n",
    "\n",
    "        # Iterate through motif_dict\n",
    "        for tf, motif in motif_dict.items():\n",
    "            motif_len = len(motif)\n",
    "\n",
    "            # Get both forward and reverse PSSMs\n",
    "            pssm_fwd = motif.pssm\n",
    "\n",
    "            # Search with forward PSSM\n",
    "            forward_matches = list(pssm_fwd.search(promoter_seq, threshold=threshold)) # Scan Promoter sequences with forward motif patterns\n",
    "            for pos, score in forward_matches: # Iterate through matches\n",
    "                start = pos \n",
    "                end = pos + motif_len\n",
    "                if start < 0 or end > len(promoter_seq):\n",
    "                    continue # Skip when start value is negative or end value is over the promoter sequence length.\n",
    "                results.append([gene_name, gene_id, tf, \"+\", start, end]) # Add matched data into empty list to write CSV file\n",
    "            \n",
    "    return results\n",
    "\n",
    "# Save results to CSV\n",
    "def write_results_to_csv(results, filename=\"motif_scan_results_threshold_1.0.csv\"): \n",
    "    with open(filename, \"w\", newline=\"\") as csvfile:\n",
    "        writer = csv.writer(csvfile)\n",
    "        writer.writerow([\n",
    "            \"Gene_Name\", \"Gene_ID\", \"TF\", \"Strand\", \"Start\", \"End\"\n",
    "        ]) # Write header row\n",
    "        writer.writerows(results) # Write the rows in CSV file\n",
    "    print(f\"Results written to {filename}\") # Print this message to confirm\n",
    "\n",
    "# Run functions\n",
    "if __name__ == \"__main__\":\n",
    "    tf_names = [\"IRF3\", \"IRF7\", \"IRF9\"] # Set the list to fetch multile transcription factors\n",
    "    fasta_file = \"/Users/sang/bioinformatics_msc_project_2025/github/TFBS-explorer/outputs/Promoter_sequences.fasta\" # Sequence data to scan\n",
    "\n",
    "    motifs_dict = fetch_motifs_from_jaspar(tf_names) # Store function result in variable\n",
    "    scan_results = scan_promoters_for_motifs(fasta_file, motifs_dict, threshold=1.0) # Store function result in variable\n",
    "    write_results_to_csv(scan_results)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7427b0a",
   "metadata": {},
   "source": [
    "# Playground\n",
    "\n",
    "- try using only forward strand and only forward profile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c4fd7f5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ifnb1 = \"GTAGTTTCTTTTTTCTTTGAGAGCCATCATCACCATCATGGTTGACACCATGAACCTATCTGAAGATGTCAGCCATAGACTGCTTGATATTCTACAGGAAAGATCACAGTTTTAAGTGCAATCTACCCATGTTATTAGCAGTGTGTATCTTTCACACATTACACAGCCTCTCTAAGCCTCATTTCTCTCCTCTGTAAGATGGGGATGATAATAACCCATCTCAAATGTTTACTATGAGGATTATTCAAAGAATGGCAAATAGCAAGTGCTTAATAAATGATAACTAGTACTACCGCCACTACTGTTGTTTTTATTGTATTAGATTATGAACTCTCTAAGGACCATTTCCGGATGGAGGATAAGAGACCATTTGATGTGGGCAGTGATGAGGCCTTCTGTTGCACCTGGAAAGGTCAACTATATACAAGCCTGCAAGTCATTCTATAGGAGCAGGCCCCAGTGACCAGACTCTATAGACTGTCTCCTCTTTCCTGAGAGGGACAGCCATCTCTAGGTTGACTAACCTCTGAAGCTCCTTGCATTGGCTTTTGTGCTATGAGCCATGGATGATTCCAGACTAATCCGAGAATGCTCGTCAAAACCCCAAGGAATTACTCAAATACTGACATAACAGACATTTTTGAGTGGAAGAGCCGAGTTTTTTTTAATATTCTGAAACTCATTGTTTTTAAAATGCATGAGATGGCCAAGGTCTTGCTAAGAGCTGGCCTGCAAAGGCCAAAAGGCCAGAGAGAATGAAACCCATAGAGAGGCAGAATAACCAGAAAGGTTGGGACTCGTTTATTTTATAATGTAAATTAGTCTATTATGAAACAATACTTGTTTACTGGTGGAAAATTGGAAAATACAAAGAATAAAAGGAGGAAAAAAATCACTCTTTAGTTTCACAAGCCAAATCAAGCCACTATTAAAATGGTGGTTTACTTCCTTTTATTAATTTTCTCTACATATTTTTGCATAATCATGTTGTATGTACAATTTTATGTTCTATTTTTCAATATTAACTGGTGTCTTTCAAATTTCCTAATGACAAAAATAATATATGCTCATAATAGAACATTTTAAATGCAAATAAAACAAAATAAATGTTAAAATTTAGTAATATTTATTAAATTTTCTCCAAGTGCACGAAATTACAAATGTAACAACCTAATTCCCTAGTGGCCTAATAACCCTATTTCCCAGACCTCTTCTCATTACAAGGAAAAACTCATATGCAGATAGTTCTAAAGGTATGAAGTGAAAAGATAAAGATTTTTCTTCCTTGCTGCATCCTCACCCCATCAGCATTATTCCCCAGGGTAACTACTATTAATAGATAGTAATTCTACCCAAAGGAAAAAATCATATGCATATAACAGCATCATATGTATACCTTTCTAGTAACTTACAAAACAAATGATAATATCATATCCTTTCTTATGTGTATTGCTCTTTTCACTAAATGTATCTGTGATATGTGTCTATATCAGCTGATTGTCCTTTTTGATGGCTGAATAATATTCCATCTTGTCCACGTGATAGTATTACTTGACAAGCTCCCTGCTGATGGACATTTGTCTTTGTTACTATGATAGTAATATAATCAACATTTATATATGTTTTGTATGTATCTATAATACACATGCACATACACATGCATATTTCTGCAGGGATAGCCATAGTAAATAACTAGTAACGGTATTGCAAGTTAAAGGAACAATCTCATTGCTTGAAATTTTAAATTTTGAAATACACTGCCAATTTTCATGGTCTCTCCTTGTAAGCTAGTTTGGGCTTTCTCACAGCATGACAGGCTCAGGGCAGTCAGACCATCCTGGCCAAAGAGCAGAGTGCCACAGACCACAACTGCTTCTAATCAGCCATCTTCCCAAAGCCTTCTCTTTTTTCTATTAATAACTTTGTATGAGATTCCATCTTAATACTTTTCTGTTGTTTGGTCTTGTAAGAGCTTATTTTTCCTGAACCAGGAAGTGGTTCAGGGCGGTTTTTCTAACTTCACAGAGCTCCCTCTTCTGTTAGCTTTTGTGAAATGGTCAAAAACATAGCAGCCTGCCTTCTGAGTTCTCCATCCCACCCTGGTTGGGCCTTCTCTATCCTTGTCTGTGTTGTTTATATCCTGCTGAAGTGTGATTCCACTTGTGGCAGTTTCTCCTCTGTGTAGGATCAAAAGGGCTGTGACTGGTTGGTTTGAAAATTTCTTATACCCTAGACTATTCCAGTGCCTTTCAGAAGTTTCCAAGGCCCTCTCACACTAATCTATTATCATATTGGGCAAAACTCCTTGCAGTTTCAGCTACTATTCCCTGATTGACTTTTCAGTAAATCTATCTCTCAGTCTTTCAGTATCCAAAGAAGATTGGTTCTAGGACCACCATCCCGCTGCCTCCACAGATACCAAAATCAGAGGATGCTCAATTCCCTCTTATAAAACGTTGCAGTATTTGCATATAATCTGCACATGTATTTCTGTATATTTTAAATCATCCCTAGATTACTTATAATACCTGATACAATATAAATGCTAAATAGCTGTAACACTGTATCTTTAAAATTTACATTATTTTTTGTTGTTGTATTATTATTTTTATTGTATTTTTAAAAAATATTTTCCATCTACAGTCAGTAGAATCCACGGATACAGAACCTATGGATAGGAAGGACCAACTGTATCTTTTAGTGTTTTGAGGTTCTTGAATTCTCAGGTCGTTTGCTTTCCTTTGCTTTCTCCCAAGTCTTGTTTTACAATTTGCTTTAGTCATTCACTGAAACTTTAAAAAACATTAGAAAACCTCACAGTTTGTAAATCTTTTTCCCTATTATATATATCATAAGATAGGAGCTTAAATAAAGAGTTTTAGAAACTACTAAAATGTAAATGACATAGGAAAACTGAAAGGGAGAAGTGAAAGTGGGAAATTCCTCTGAATAGAGAGAGGACCATCTCATATAAATAGGCCATACCCATGGAGAAAGGAC\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "3335550d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ifit1 = 'GCTGATAAGAGATTTTAGACTACAAGAGATGACCCTGGAAAAGACAGGGGAAATTTCTATGGAGCCCATTCCTTAATAAGTTGTGATTGGATATTTGAGACCCTTCAGCAGAAATGGCTTAGCATTTGTCCATATTATCATATTATAATTTTTAAAAAATACACTCAGAATAACATTCTTTTTCCATATCATCCTGTGAATGAAGAAGAGACTAAGGAATTCATTACCCATTGTGTAGGTGAGGAAACAAGGCATTGAGTCGTAAAATACTGAATCTGGGCCTCACAGAACAGGAAGTGGGTCCAGGGATTGAGTCCCATATCTCTCAGTCAGAGAATTTCAGACATACAATAGACATTAGGGATAATTGAGGCCCTGGTTCCCGGCTCAGGGCGCAGGTCTACATCAAAATCACCTGAAGATTTCTTTCTTTTTTTTTTTTTTTTTTTTTTTGAGATGGAGTTTCACTCTTGTTGCCCAGGCTGGAGTGCAATGAAATGGCACGATCTCGGCTCACCACAACCTCCACCTCCCGGGTTCAAGCGATTCTCCTGCCTCAGCCTCCCGAGTAGCTGGGATTACAGGCATGTGCCACCACGCCCGGCTAATTATGTATTTTTAGTAGAGACGGGGTTTCTCCATGTTAGTCAGGCTGGTCTCGAACTCCCAACCTCAGGTGATCCGCCCGTCTTGTCCTCCCAAAATGCTGGGATTACAGGCGTGAGCCACTGTGCCCAGGCTGAAGATTTCTTTTTAATACTGAGCCAGAGGCCTGTACCTAAGTATATGTTCAATTGTTCTGGAATGGGCATAGTTAGCATTATTTTTGAGATTTACCTGGCTGATTCTAATGTGCAGTGAGGGTTGAGATCTGCCAACCTATTCCAAATGACTTCTCAGCTGAGGAAATTGAGATTCATAGAGGTGAAGCAGTTTATCATAACTGGTCAGTCCCTTGCAGAGCCCAGCTTAGACTCCTGGGAATGAACTCACAACCTAGTGCTGAAATTTCATTTTCTAAATGTTTGATCCTCTCTGGTTCCCATTTTCTTTTTTATCTGTAAACACACACACACACGCGCGCGATTGGCTTTTCTTTCCTTAATGCTGTTCTATCCTGTGTCCCTCATGGGTTTGTTTAGCCCAAAGAGACACTAAGCAAGATTCTGAATACCTAATCAGCCTGGGCCATAGCCCCGTAGAACCAAGACTGCTGCCAAATTCACAGCAATGGCAAGATAAACTTTTTGTTCATATCTCTGCATGCTATCTAGTGCATTGGCAAATTGAAAACCCATTTTTTCCCATAAGAATTCCAAAGACATCTTGCTCCCCAAAAGATGTTTTCTAATTCATCTGTTTCAGCATTTGCTTATGGCCTGATCATGTATCTGAGTGCAAGACAAAACCTGTCCTATTAATTCAGAAATTCTTATTTGTGTAGAACTTTCTAACTAGTTGTTGAACTCTCTGTGGAAGCTTACAGAATTGGTGAGAGCTGACATGGGAGATGAGTCTGGCAGGGAGTTGAGCAGAAACTGTTTAGATCTGGCAGTGTGCCCAGTTAAGTGTTTCCAAAAGGAAAATACTCTTGCACCTGCTCCTCTTCTGCTTCTTATGTTAAAAGACAAAGTAGGTAGGTTTTTTTAAAGGAAACATTGTTTTTATATATTGTTATGATTATTTGAGACAGGGGCTCACTCTGTTGCCCAGGCTGGAGTGCAGTGGTGTGACCATGGCTCACTGCAGCCTCTACCTCCTGGGCTCAAGTGAGCCTCCCACCTCAGCCTCCTGAGTAGCTGAGACCACAGACACACACCACCACACCCAGCTTTTTTTTTCAGTTTTTTTGTAGAGTTGGGGTCTCCCTATGTTGTCCAGGCTGGTCTACAATTCCTGGGCTCAAGTGATCCTCCCGCCTTGCCTTCCAAAATACTGGGACTTACAGGCCTGAGCCAATGCACCCGGACAACAATTTTTTATTTTTTATTTATTTATTTTTTTAGACGGAGTCTCGCTCTGTCACCAGACTGGAGTACAGTGGTGTGATCTCGGCTCACTGCAACCTCTGCCTCCCAGGTTCAAGCAATTCCCCTGCCTCAGCCTCTCGAGTAGGTGGGACTACAGGTGCACACCACCACACCCAGCTAATTTTTTGTATTTTAGTAGAGAGGGGGTTTCACCATGTTGGCCACGATGGTCTCCATCTCCTGACCTTGTCATCCGCCCACCTTGGCCTCCCAAAGTGCTGGGACTACAGGCATGAGCCACCGCACCCAGCCAAGAATCATTATTTTTAACTTGATGACTGAAAATAATAATAATAATAGTTACCACTTATTTGCATGCTTCTATGTGCCAGGTAGTTGCTAACTATTTAAACTCAAATTCCATGAACTGTAGTGGAGGTTGTACTGGAATTTGATTCAGAATGACAGTGTCCATGATGGAGCAATAGAGGGCTCTCTATTTCAAACCATACCTCCTTGCTTTTACCTCCTGCCTAAGTCATCAGGGGTTAGAAGGCTTTCTAGGTATTGGTCTCTTTCCTTCATTCCTAAACCAGATTGGTTGCTTATTTCCGTCAAGCTGAAACCAAAAGTAAGCAACCAAAAAGCAACCAGCAACCAAAAGCCTTGTTACTCAATTAATTAAGAGTAGATTTTTATATTTGATAGTAGGTTCCTTCTAAATATAGAAACTGAAAATAGAGCTATCTCCTTCAATTCTCCTTTTTCTGTGTATTCATCCAGAATCCAGCCACCAACTGCCACAATAGGCAGCAATGGACTGATGTTCTTTAGGGAGGACGTGAATCTCGTTCCAAATGCTGGCCAGTCATTGGGTTTCTGCAGCACTAGAAACATCTATGGTTGCAGGTCTGCAGTTTATCTGTTTTAAAATAGAAACAAAGTTTCATTCCCCACCCCCCCCCGTCAGCAGGAATTCCGCTAGCTTTAGTTTCACTTTCCCCTTTCGGTTTCCCTAGGTTTCCAACTT'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "4bfe6d5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "MX1 = 'AATGAAATAATGGCATTCACAGCAACCTGGATGGAGTTGGAGACCATTATTCTAAGTGAAGTAACTCGGGAATAAAAACCAAACATTGTATATTCTCACTCATAAGTGAGATTATACAAGCTATGAGGATGCAAAGTCATAAGAATGATACAATAAACATTGGGAACTCAGGGGGAAGGGTGAGGGTGGGGTGAGGGATAAAAGACTAGACATCGGGTACAGTATACACTGCTTGGGTGATGGGTGCACCAAAATCCCAGAAATCACGATTAAAGAACTTGTCAATGTAACCAAACACCACCTGTTCCTCAAAAACTATTGAAATAAAAAATAAATACATAAAAGAAGTTGAGGTTAAATTAATCTAAGTAGGGAGTGCATTTGGGCCAAGCTTGAGGATTGCAACGTTGGAGCACAGATTTAAGTTGCCCTGAATCTACACTCTGATAAGTGGCATTTATAAGTGGATTTTTAAAGGCAAAAGGGGGAGACAGGGAGAGGGCTGATACAAAGCTGTTTGTCAGGAATTCTTATTGGTTTACAGAAATAACATTGATAAGTGCTTATTAGTCCATTTTCATGCTGCTGATAAAGACATACCTCAGACTGGGCAATTTACAAAAGAAAGAGGTTTATTGGACCTATAGTTCCACATGGCTGAGGAGGCCTCACAATCATGGTGGAAGGCAAGGAGGAGCAAGTCACATCTTACATGGATGACAGCAGGAAAAAAGAGAGAGCTTGTGCAGAGAAACTCCCTTTTTTAAAACCATCAGATCTTGTGAGACCCGTTCAGGATCATGAGAACAGCACAGGAAAGACCCACCCCCAAGATTCAATCATCTCCCACCGGGTCTCTCCCACAACATGAGTGAATTATGGGAGCTACAGGATGAGATTTGGGTGAGGACACAGAGCCAAACCATATCAGAGTGATTGACTATACATTGTTAAGCTACAGGGTGTGGATTACAGCGTCCAGTGTGGCATTATTAGGTTAATCCATAGCTATTAATACCTGTGGCAATAGTAAGAGGTTTCAAGAGATGAACACACAGCTGAAAGTGGGGAGTAGGGAGTGGTTGCTGTCTTATTTTAATGTCTCTCTGGGCCTGATAATTTTAAAAGCCTTGCATTCTTCGGATACAAGTTCCTTTCTTTTTTTCAAGAACAACAGAGAAAGGCTGCCTGGCTACCTTTACCTGCAGCACCTCTGATAGTGGCTGCTTGGCAGCTGGTTAGGTAAACCCAATTTCTACTGGATAAACTTCATGGACATTCACCCATGTTAGTACTGGGGAATGCTCTGAAATAGCAAACTTGTCCATACAGCTCCCTTGCTGCTCAGAGGGTCTGGAGAGGAACAGCAGAGGCCCAGCCCCAAATACTTCTTGTACCGGCTTCTATTATGAGAGGTGATGGCACCTTTAGGCACTTATATTTCAGGTAGAAGGAGCAGGAATAAAGAAGAGAGGGCCACAATCTGAAATATCAGTTCTGGATCTCCCTGGGGGCAAGACACTGGCCCTTAGCTGAGCTATTCTGTCATCACTGTGAGCCCTCATGCTCGACCTCACCGGAGCACCCCTCGATCATGTGCTGAATGCCTAGCACGCGCCTGGCACTGTGGGGTACAGAAATCTAGAGCACAGAGCTGAGTTTTCAGGGCTTAGCTGTCCTCACATCTGTGACACAAACAAGGATCAAGGAAAGTTTTGTTTTTGTTATTTTTATTTTTTTTTAAGTTGGGGCCTTACTCTGTAGCCCAAGCTAGAGTGCAGTGGTGCAAACATAGCTCATTGCACCCTCGAACTCCTGGGCTCAAGTCATCCTCCCACCTTGGCCTCCAGAGTAGCTGGAAGCCCAGCTCAAGGAAAGCATTAACATGTTCTACTTTTCACCATAACTCCTGTGTATTGGGAAATGGAAAATGAGCTTTTGAAGTGGCCTATTCGCTTCCTCCATGAGTACTTCCTGAACACTTTCCTCTTGGCCACTGCTCCCAAGACCCCCTCAAAGTCTCTCTCCTGGACTTCCAGCTTCTGGAATGAGCCCCAAGAGTGGCACTTAGCTCCAGCTCCTAGTAGCTCACTCCTTCCTTGCTCACCTGTGCCCCCCACCTGTCTCCTTAGGGGCAGTGGCTTTGTCTTACTCATTTGCATATCACAAATGTCCAGCATAGAAGCCAGCATGGGGCCCAGTAAGTGTTTGTGGAGAAAGAAAGGGAAGAGGTGAAGAAGGGAGGAGACGGAGGGAGAAAAGGAGGGCAAGAGAGGGGATGTCATGTCATGTCATGTCATGTCATGTCAGTTCTCCCTTCCTAGATCTGGACAAGGGAAAGAAAAGCAAGCAAGCCTGACCGCCTCAGTGCAGACCCTCTAAGATGCAAGCCTTCCCCAGGTAAGACAGTTCCGCAGGGTGCTTCTGTGGGCTGGCTTTCTGAGCAGCCATCTCAAAGTATGCGAAAGAAATAGATTTTGGGCCGGGCACGGTGGCTCATGCCTGTAATCCCAGCACTTTAGGAGGCCGAGGCGTGTGGATCACCTGAAGTCAGGAGTTCAAGACCAGCCTGGCCAACATGGTGAGACCTTGTATCTACTAAAAATACAAAAACTAGCCGGGCTTGGTGGCGCATGCCTGTAGTCCCAGCTACTCAGGAGGCTGAGGCAGGAGAATCGCTTGAACCCAGGAGGTAAATGTTGCAGTAAGCCGAGATTGCCCCACTGCACTCCAGCCTGGATGACAAGAGTGAGACTCCATCTCAAAAAAAAAAAAAAAAAAAGATTTTAGGGGTAAAATATTTTGCTTTCCTTCAGTGGCTGGGCCACCTGTACACAATGCCTGAACCCAGGTCCTGCAGCCCCCAGCTTGGCTCTCCAAAGCTCACCAGTATCAAAGGATTTCAGCTTCTGCTCCTCTCCGCAGGGCTCTCTGGTGGGCCCCAGACAGAGATGAAGGAAAAGTCTGGCCTTGACCTTGAGGACCAAAAGCGACCACGTC'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "b0663b3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "TLR7 = 'ATGGGCGTGGGACCCTCCGAGCCAGGCGCGGGATATAATCTCCTGGTGTGCCATTTGCTAAGACCATTGGAAAAGCACAGTATTAGGGTGGGAGTGACCCGATTTTCCAGGTGCTGTGTGTCACAGCTTTGCTTGGCTATGAAAGGGAATTCCCTGACCCCTTGTGCTTCCCAGGTGAGGCAATGCCTCGCCCTGCTTTGGCTCATGCTCGGTGTGCTGCACCCACTGTCCTGCGCCCACTGTCTGACAAGCCCCAGTGAGATGAACCCAGTACCTCAGTTGGAAATGCAGAAATCACCCATCTTCTGCATTGCTCACGCTGGGAGCTGTAGACTGGAGCTGTTCCTATTCAGCCATCTTGGAACCACCCTACAGTTTCTTTCAATACTTTTCATGTTCAATGTCTTACACTTAAATTTTTCAACTACATGAAATTTATTTTGATATAAGAAATGAGGTTGGGCCAGGCGCAGTGGCTCACACCTGTTATCCCAAAACTTTGGGAGGCTGAAGTGGGCGAATCATGAGATCAGGAGTTCAAAACCAGCCTGGCCAATATAGTGAAACCCCACCTCTACAGAAAATACAAAAAATTAGCCGAGCGTGGTGGCGGGTGCCTGTAATCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCTGGGAGGTGGAGGTTGCAGTGAGCAGAGATTGCGCCACTGCACTCCAGCCTGGGTGACAGAGTGAGACTACATCTCAAAAAATAAATAAAAAGACAAATGAGGTTGGACAATAGCTTGATTTTTTAAAAACTCCAAGCCAGTTGCCCAATAATCCATTTTGCCATTGATGTGACTTGAAACTAAACACCCATATGTGCTTTTACTTAATTCTGCTCACTTCTCAGTTTCACTGATGCACTGGTACTATGCCATTGGAGGCAATACAGCATTCTAATGCATTTACAGATATGGAAGAGAAATCCTCCCCTTTTATTTTCATTCAGAAATTGCCTGGCTATTCTCACATGTGGTCTTACCCCAGTCAGACCCAACACCTCACTTTTATAACAAATATTTGGTAATGCACCCTTTGTTATATGAAAGGAGATATTTGTGGATAATGTAACCCCAGTCTTCATGATAAACAAAAAGGCCCAGCTGATTTCCAAAATGCACCCCAGTTTAGAATCAGTCTGGCAAGTATCACATGAAATCCTATTGGTATTTGATTGGGATCACACAGCATTTGCAAATCAATTTAAGCAACATTTCTATTTTTACAATATTGTGGCCTCTAGCCCCAATAACAATTATTTCTCTTCATTTACTTATATCTTTGTGTCTGGGCAGGGTCCTGCCAGGAAACAGACGGCATGTTAAAGTGAGAAACTGATGAGTTCAGCAAAGTGACTATTTATATATTTGGACAGGATTTAAGGAAGTAAGAAAGGATGATGCAACACTTCAGAGGGGAGTCTTTCCGACCTCAGGCTGAAGGAGAAGGAACGATTACTGGAATTCAGGGAGGAGAGCATCACCAAACAAGAGCTTCATTAGAGGACTGCAGCCAACGCAGGGCCAGGCGGAGGGAGCCAGAGGGAGGCAGGCTCTGCTCTCCCTCTTCCTGCCCTTCAGTTTCCAACCACGGCCTCCTATTGGCCAAACCCAACCAGAAGCCAGCGAGCAAGGGGCTACTGATGAAGCACATATAGCTCAGCCTCCAGAGACACAGAACAGGATAAAAGGCTGAGACAGTGGGTCTGGTGGGGCAAAGAGAAAGCTTGCACCTGCCAGGTAAAGCATTATAGTCCCCATCCTCCCCCCACCACCTTAGTTCTTGTGCATTTCCCATCAGTTTTCTTCCAAAGCATTTCAGATCTCACTGATTTGAATGGGACCTCTTCTTCTATTACCAATTTGGATCAGTAATTATTTATGTATGAGAAAACTATTGATTTTTACATATCGTTGCATAGCAGAGTGGTTTAATAAGGAGACATTTGGTTTTTACTGCCTGTGTGGGTATCACTTGCTATGTGACTTGAGGCAAATCCAATATTTCTTCTGTTATAAATTCCAGTATTTGTAAAAAATGGGTAATAAGATCTCTATTTTATATAGTTTTAGTATTTAATGAGATAATACATATAAAGTCATTAAAACAGTGTCTGGCTCATAAAAAACCCTCAATAAATGTCACTTATTACTGTATCTGGTTTTTGAGCTGCTCTATTGCACCATTGAGTTTTCAGCCCAGTATATGTTAACCCTGATCATTATCTGCAGAAGTCCCCGTGCCACACTCTACATCATCCAAATTCTCTCCAGGTGGACTAAGTAGATTAAAGAACTTTAAACATAACTACCATATTTTGGCTCTATCTACAAAATGTCCAATAATCAGTTAAGAAAGGAACAATTCTCTTGGGGCCCACACTTTGAGAAGCAAATGCAGCTGAACTTTTTTAGAGGAAAGTGAGTGAACCAACTGGTAGCTTTGCCACTGCTTAAAAACCAGCATCCTTTCCAGCTGGGTCTAAGACAGAATAAGGTAAATTTAGATATGTCTCTAATATATCTATAGAACAGTGGTTCTCAACCCGGGGTGTTTTTGCCCCTTAGGGGATAATTTGCAATGTCTGGAGACATCTGTGATTGTCATAACTGGAAGGGGGCAGTGCTATTGGCATCTAGTGGGTATAGAGCAAGGGTGCTACCAAATATCCTATGGTGCAACAGAGAATTATCTGGTCAAAAATGTAAATAGTGCTGAGGGTGAGAAACCCTGCTATAAAAACGAAAGAAATTTGGTCTACAGAGTTGTTTGGATTTAGACAAGACGTTGCCCCAATAGTGGTGATAGAAATAAGAGGAACCCCGTGCTTTTGCAAAGCCCATATCTGGGGTGGCTTAAATAATCATGCTCCTCCCCATCCCCCGACCTGATCTTTGTAGTTGGAAACTCCAGGGCTGGCTGCCTGTAGTCTTTGTGACTACACTTCCTGCCTCCCATC'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d0bc18a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "positions = [751,768]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2a1e2f65",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GAGAATGAAACCCATAG'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start, stop = positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "93df6e2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GAGAATGAAACCCATAG'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ifnb1[start:stop]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f9bfb9f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ifnb1[start:stop])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "19305523",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "\n",
    "ifnb1_dna = Seq(ifnb1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3a53bb0c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('GTAGTTTCTTTTTTCTTTGAGAGCCATCATCACCATCATGGTTGACACCATGAA...GAC')"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ifnb1_dna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7b53d744",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('GAGAATGAAACCCATAG')"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ifnb1_dna[3000-2249:3000-2232]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2157eb48",
   "metadata": {},
   "outputs": [],
   "source": [
    "ifnb1_dna_reverse = ifnb1_dna.reverse_complement()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "eb2ad041",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('GTCCTTTCTCCATGGGTATGGCCTATTTATATGAGATGGTCCTCTCTCTATTCA...TAC')"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ifnb1_dna_reverse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "25288a7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "positions2 = [2232,2249]\n",
    "start2, end2 = positions2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "bc98a3eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('CTATGGGTTTCATTCTC')"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ifnb1_dna_reverse[start2:end2]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53c91afd",
   "metadata": {},
   "source": [
    "## Threshold 3.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "72cde68a",
   "metadata": {},
   "outputs": [],
   "source": [
    "match1 = [751, 768] # IRF3\n",
    "match2 = [2903, 2920] # IRF3\n",
    "match3 = [2909, 2926] # IRF3\n",
    "match4 = [1092, 1105] # IRF7\n",
    "match5 = [1097, 1110] # IRF7\n",
    "match6 = [1149, 1162] # IRF7\n",
    "match7 = [2694, 2709] # IRF9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "cea0726e",
   "metadata": {},
   "outputs": [],
   "source": [
    "start1, end1 = match1 # IRF3\n",
    "start2, end2 = match2 # IRF3\n",
    "start3, end3 = match3 # IRF3\n",
    "start4, end4 = match4 # IRF7\n",
    "start5, end5 = match5 # IRF7\n",
    "start6, end6 = match6 # IRF7\n",
    "start7, end7 = match7 # IRF9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f02831c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GAGAATGAAACCCATAG\n",
      "CATAGGAAAACTGAAAG\n",
      "AAAACTGAAAGGGAGAA\n",
      "ATAAAACAAAATA\n",
      "ACAAAATAAATGT\n",
      "CGAAATTACAAAT\n",
      "ATAGAAACTGAAAAT\n"
     ]
    }
   ],
   "source": [
    "# threshold 3.0\n",
    "print(ifnb1[start1:end1]) # IFNB1\n",
    "print(ifnb1[start2:end2]) # IFNB1\n",
    "print(ifnb1[start3:end3]) # IFNB1\n",
    "print(ifnb1[start4:end4]) # IFNB1\n",
    "print(ifnb1[start5:end5]) # IFNB1\n",
    "print(ifnb1[start6:end6]) # IFNB1\n",
    "print(ifit1[start7:end7]) # IFIT1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a39974e",
   "metadata": {},
   "source": [
    "## Threshold 5.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "3a878273",
   "metadata": {},
   "outputs": [],
   "source": [
    "match1_t5 = [2903, 2920] # IRF3 in IFNB1\n",
    "match2_t5 = [1149, 1162] # IRF7 in IFNB1\n",
    "match3_t5 = [2585, 2602] # IRF3 in IFIT1\n",
    "match4_t5 = [2691, 2708] # IRF7 in IFIT1\n",
    "match5_t5 = [1281,1294] # IRF7 in IFIT1\n",
    "match6_t5 = [2694,2709] # IRF9 in IFIT1\n",
    "match7_t5 = [2738,2751] # IRF7 in TLR7\n",
    "match8_t5 = [2777,2790] # IRF7 in TLR7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cda7f59d",
   "metadata": {},
   "outputs": [],
   "source": [
    "t5_start1, t5_end1 = match1_t5 # IRF3 in IFNB1\n",
    "t5_start2, t5_end2 = match2_t5 # IRF7 in IFNB1\n",
    "t5_start3, t5_end3 = match3_t5 # IRF3 in IFIT1\n",
    "t5_start4, t5_end4 = match4_t5 # IRF7 in IFIT1\n",
    "t5_start5, t5_end5 = match5_t5 # IRF7 in IFIT1\n",
    "t5_start6, t5_end6 = match6_t5 # IRF9 in IFIT1\n",
    "t5_start7, t5_end7 = match7_t5 # IRF7 in TLR7\n",
    "t5_start8, t5_end8 = match8_t5 # IRF7 in TLR7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80940e3e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CATAGGAAAACTGAAAG\n",
      "CGAAATTACAAAT\n",
      "CAAGCTGAAACCAAAAG\n",
      "AATATAGAAACTGAAAA\n",
      "GCAAATTGAAAAC\n",
      "ATAGAAACTGAAAAT\n",
      "CAAAAATGTAAAT\n",
      "TAAAAACGAAAGA\n"
     ]
    }
   ],
   "source": [
    "print(ifnb1[t5_start1:t5_end1]) # IFNB1 - IRF3\n",
    "print(ifnb1[t5_start2:t5_end2]) # IFNB1 - IRF7\n",
    "print(ifit1[t5_start3:t5_end3]) # IFIT1 - IRF3\n",
    "print(ifit1[t5_start4:t5_end4]) # IFIT1 - IRF7\n",
    "print(ifit1[t5_start5:t5_end5]) # IFIT1 - IRF7\n",
    "print(ifit1[t5_start6:t5_end6]) # IFIT1 - IRF9\n",
    "print(TLR7[t5_start7: t5_end7]) # TLR7 - IRF7\n",
    "print(TLR7[t5_start8: t5_end8]) # TLR7 - IRF7"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5fc94729",
   "metadata": {},
   "source": [
    "## Threshold 7.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "e52279c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "match1_t7 = [2903,2920] # IRF3 - IFNB1\n",
    "match2_t7 = [2888,2901] # IRF7 - IFNB1\n",
    "match3_t7 = [2691,2708] # IRF3 - IFIT1\n",
    "match4_t7 = [1574,1587] # IRF7 - IFIT1\n",
    "match5_t7 = [2590,2603] # IRF7 - IFIT1\n",
    "match6_t7 = [2696,2709] # IRF7 - IFIT1\n",
    "match7_t7 = [2694,2709] # IRF9 - IFIT1\n",
    "match8_t7 = [2738,2751] # IRF7 - TLR7\n",
    "match9_t7 = [2777,2790] # IRF7 - TLR7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "c9af0ea9",
   "metadata": {},
   "outputs": [],
   "source": [
    "t7_start1, t7_end1 = match1_t7\n",
    "t7_start2, t7_end2 = match2_t7\n",
    "t7_start3, t7_end3 = match3_t7\n",
    "t7_start4, t7_end4 = match4_t7\n",
    "t7_start5, t7_end5 = match5_t7\n",
    "t7_start6, t7_end6 = match6_t7\n",
    "t7_start7, t7_end7 = match7_t7\n",
    "t7_start8, t7_end8 = match8_t7\n",
    "t7_start9, t7_end9 = match9_t7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "7a9beeea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CATAGGAAAACTGAAAG\n",
      "CTAAAATGTAAAT\n",
      "AATATAGAAACTGAAAA\n",
      "CCAAAAGGAAAAT\n",
      "TGAAACCAAAAGT\n",
      "AGAAACTGAAAAT\n",
      "ATAGAAACTGAAAAT\n",
      "CAAAAATGTAAAT\n",
      "TAAAAACGAAAGA\n"
     ]
    }
   ],
   "source": [
    "print(ifnb1[t7_start1:t7_end1]) # IRF3 - IFNB1\n",
    "print(ifnb1[t7_start2:t7_end2]) # IRF7 - IFNB1\n",
    "print(ifit1[t7_start3:t7_end3]) # IRF3 - IFIT1\n",
    "print(ifit1[t7_start4:t7_end4]) # IRF7 - IFIT1\n",
    "print(ifit1[t7_start5:t7_end5]) # IRF7 - IFIT1\n",
    "print(ifit1[t7_start6:t7_end6]) # IRF7 - IFIT1\n",
    "print(ifit1[t7_start7:t7_end7]) # IRF9 - IFIT1\n",
    "print(TLR7[t7_start8:t7_end8]) # IRF7 - TLR7\n",
    "print(TLR7[t7_start9:t7_end9]) # IRF7 - TLR7"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "141e07f1",
   "metadata": {},
   "source": [
    "## Threshold 10.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "98eb685e",
   "metadata": {},
   "outputs": [],
   "source": [
    "match1_t10 = [2590,2603] # IRF7 - IFIT1\n",
    "match2_t10 = [2696,2709] # IRF7 - IFIT1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "6eb6b982",
   "metadata": {},
   "outputs": [],
   "source": [
    "t10_start1, t10_end1 = match1_t10\n",
    "t10_start2, t10_end2 = match2_t10\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "8d461d24",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TGAAACCAAAAGT\n",
      "AGAAACTGAAAAT\n"
     ]
    }
   ],
   "source": [
    "print(ifit1[t10_start1:t10_end1])\n",
    "print(ifit1[t10_start2:t10_end2])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
